---
title: "3_BarPlots"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
#install.packages("BiocManager")
#BiocManager::install("Seurat")
#BiocManager::install("ggplot2")
#BiocManager::install("sctransform")

library(Seurat)
library(ggplot2)
library(dplyr)
library(ape)
library(cowplot)
library(Matrix)
library(EnhancedVolcano)
library(knitr)
library(readr)
library(RColorBrewer)

#Specify your preferred directory for all input + output
dir = "/Users/nathanewing-crystal/Ari Molofsky Lab Dropbox/Ari Molofsky Lab Team Folder/Nathan Ewing-Crystal/RNASeq/Lesion snRNASeq 3 Cxcl12/seurat"
#Specify exactly where your seurat files live
datafolder = "Data/Seurat"

project<-"cxcl12_lesion_snRNASeq"  

#set up folders
Plotfolder = "Plots"
load(file.choose())
```

Barplot for any two (or more) categories

```{r}
#Pick metadata columns
clustercolumn = "celltype"
samplecolumn = "genotype"
#pick a reasonable number of cells per sample to normalize by
ncells = 2000 
cols = brewer.pal(n = 6, name = "Dark2")
cols.4 = brewer.pal(n=4, name = "Dark2")
cols.3 = brewer.pal(n=3, name = "Dark2")
cols.2 = c("#1B9E77", "#D95F02")

#If you want to only compare particular samples/conditions, split further by another metadata column: 
# split = NULL
```

#MYELOID (with SAM subclusters [from 4_Myeloid.Rmd])
```{r}
##### 1 ####

# Below uses total sobject
# measure myeloid subsets across timepoints as % of total cells
sobject_sub = sobject
clustercolumn = "immune.subcluster"
  
  #Make a table and normalize
  r = table(sobject_sub[[clustercolumn]][,1],sobject_sub[[samplecolumn]][,1])
    
    t = r  
    #remove any clusters that don't have cells
    t = t[rowSums(t)>0,]
    
    #normalize by sample
    t = apply(t,MARGIN = 2,function(x)x/sum(x))
    t = round(t*ncells,0)
    
    #convert to percents for each *sample*
    t = apply(t,MARGIN = 2,function(x)x/sum(x))
    t = round(t*100,2)
    
    #reorder (by condition)
    tdf = as.data.frame(t)
    tdf = tdf[,c("WT", "cKO_Cxcl12")]
    t = tdf
    
    clusters_to_keep = c("Myeloid.sub_Microglia/DAM", "Myeloid.sub_SAM_0", "Myeloid.sub_SAM_1", "Myeloid.sub_SAM_2", "Myeloid.sub_SAM_3", "Myeloid.sub_PVM/BAM", "Myeloid.sub_Monocytes_DCs_1", "Myeloid.sub_Monocytes_DCs_2", "Myeloid.sub_ECM_Mac")
    t = t[clusters_to_keep,]

    name = paste0(project, "_", clustercolumn, "_cluster-frequencies_of_total")
    
    setEPS()
    postscript(file.path(dir,"Plots/Barplots",paste0(name,"barplot.eps")))
    barplot(t(t), main="Sample composition by cluster",
         ylab = "% of sample", col=cols.2,axisnames = T,
          width = .2, cex.names = 0.7,axis.lty = 1, las=2, mgp=c(), beside=TRUE)
    title(xlab = "Cluster", line = 4)
    legend("topright", legend=rownames(t(t)), fill=cols.3, bg="white", inset=c(0.02, 0.02))
    dev.off()
    
    #for pie charts
    
    t = r  
    #remove any clusters that don't have cells
    t = t[rowSums(t)>0,]
    
    #normalize by sample
    t = apply(t,MARGIN = 2,function(x)x/sum(x))
    t = round(t*ncells,0)
    
    #convert to percents for each cluster
    t = apply(t,MARGIN = 1,function(x)x/sum(x))
    t = round(t*100,2)
    
    #reorder (by condition)
    tdf = as.data.frame(t)
    tdf = tdf[c("WT", "cKO"),]
    t = tdf
    
     clusters_to_keep = c("Myeloid.sub_Microglia/DAM", "Myeloid.sub_SAM_0", "Myeloid.sub_SAM_1", "Myeloid.sub_SAM_2", "Myeloid.sub_SAM_3", "Myeloid.sub_PVM/BAM", "Myeloid.sub_Monocytes_DCs_1", "Myeloid.sub_Monocytes_DCs_2", "Myeloid.sub_ECM_Mac")
    t = t[,clusters_to_keep]
```

#T
```{r}
##### 1 ####

# Below uses total sobject
# measure myeloid subsets across timepoints as % of total cells
sobject_sub = sobject
clustercolumn = "ttype"
  
  #Make a table and normalize
  r = table(sobject_sub[[clustercolumn]][,1],sobject_sub[[samplecolumn]][,1])
    
    t = r  
    #remove any clusters that don't have cells
    t = t[rowSums(t)>0,]
    
    #normalize by sample
    t = apply(t,MARGIN = 2,function(x)x/sum(x))
    t = round(t*ncells,0)
    
    #convert to percents for each *sample*
    t = apply(t,MARGIN = 2,function(x)x/sum(x))
    t = round(t*100,2)
    
    #reorder (by condition)
    tdf = as.data.frame(t)
    tdf = tdf[,c("WT", "cKO_Cxcl12")]
    t = tdf
    
    clusters_to_keep = c("T_cell_abT")
    t = t[clusters_to_keep,]

    name = paste0(project, "_", clustercolumn, "_cluster-frequencies_of_total")
    
    setEPS()
    postscript(file.path(dir,"Plots/Barplots",paste0(name,"barplot.eps")))
    barplot(t(t), main="Sample composition by cluster",
         ylab = "% of sample", col=cols.2,axisnames = T,
          width = .2, cex.names = 0.7,axis.lty = 1, las=2, mgp=c(), beside=TRUE)
    title(xlab = "Cluster", line = 4)
    legend("topright", legend=rownames(t(t)), fill=cols.3, bg="white", inset=c(0.02, 0.02))
    dev.off()
    
     #for pie charts
    
    t = r  
    #remove any clusters that don't have cells
    t = t[rowSums(t)>0,]
    
    #normalize by sample
    t = apply(t,MARGIN = 2,function(x)x/sum(x))
    t = round(t*ncells,0)
    
    #convert to percents for each cluster
    t = apply(t,MARGIN = 1,function(x)x/sum(x))
    t = round(t*100,2)
    
    #reorder (by condition)
    tdf = as.data.frame(t)
    tdf = tdf[c("WT", "cKO"),]
    t = tdf
    
    clusters_to_keep = c("T_cell_abT")
    t = t[,clusters_to_keep]
```


#Fib
```{r}
##### 1 ####

# Below uses total sobject
# measure myeloid subsets across timepoints as % of total cells
sobject_sub = sobject
clustercolumn = "fibtype"
  
  #Make a table and normalize
  r = table(sobject_sub[[clustercolumn]][,1],sobject_sub[[samplecolumn]][,1])
    
    t = r  
    #remove any clusters that don't have cells
    t = t[rowSums(t)>0,]
    
    #normalize by sample
    t = apply(t,MARGIN = 2,function(x)x/sum(x))
    t = round(t*ncells,0)
    
    #convert to percents for each *sample*
    t = apply(t,MARGIN = 2,function(x)x/sum(x))
    t = round(t*100,2)
    
    #reorder (by condition)
    tdf = as.data.frame(t)
    tdf = tdf[,c("WT", "cKO_Cxcl12")]
    t = tdf
    
    clusters_to_keep = c("Fibroblast_Myofibroblast_Cxcl12", "Fibroblast_Late_inner", "Fibroblast_Lymphocyte_interactive", "Fibroblast_Pial", "Fibroblast_Arachnoid_1", "Fibroblast_Arachnoid_2", "Fibroblast_Altered_dural_1", "Fibroblast_Altered_dural_2", "Fibroblast_Osteoblast_1", "Fibroblast_Osteoblast_2")
    t = t[clusters_to_keep,]

    name = paste0(project, "_", clustercolumn, "_cluster-frequencies_of_total")
    
    setEPS()
    postscript(file.path(dir,"Plots/Barplots",paste0(name,"barplot.eps")))
    barplot(t(t), main="Sample composition by cluster",
         ylab = "% of sample", col=cols.2,axisnames = T,
          width = .2, cex.names = 0.7,axis.lty = 1, las=2, mgp=c(), beside=TRUE)
    title(xlab = "Cluster", line = 4)
    legend("topright", legend=rownames(t(t)), fill=cols.3, bg="white", inset=c(0.02, 0.02))
    dev.off()
    
    #for pie charts
    
    t = r  
    #remove any clusters that don't have cells
    t = t[rowSums(t)>0,]
    
    #normalize by sample
    t = apply(t,MARGIN = 2,function(x)x/sum(x))
    t = round(t*ncells,0)
    
    #convert to percents for each cluster
    t = apply(t,MARGIN = 1,function(x)x/sum(x))
    t = round(t*100,2)
    
    #reorder (by condition)
    tdf = as.data.frame(t)
    tdf = tdf[c("WT", "cKO"),]
    t = tdf
    
    clusters_to_keep = c("Fibroblast_Myofibroblast_Cxcl12", "Fibroblast_Late_inner", "Fibroblast_Lymphocyte_interactive", "Fibroblast_Pial", "Fibroblast_Arachnoid_1", "Fibroblast_Arachnoid_2", "Fibroblast_Altered_dural_1", "Fibroblast_Altered_dural_2", "Fibroblast_Osteoblast_1", "Fibroblast_Osteoblast_2")
    t = t[,clusters_to_keep]
```
