---
title: "1_qc"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

```{r}
install.packages("BiocManager")
BiocManager::install("Seurat")
BiocManager::install("ggplot2")
install.packages("dplr")
install.packages("ape")
install.packages("cowplot")
install.packages("Matrix")

library(Seurat)
library(ggplot2)
library(dplyr)
library(ape)
library(cowplot)
library(Matrix)
```
***BELOW CODE PERFORMED ON PRESIDIO***

*Edit the following code every time* and make sure the folders "QC" and ... are present in the "dir" folder above
```{r}
#Specify your preferred directory for all input + output
dir= "/mnt/DATA/molofsky_lab/newing-crystal/Lesion_snRNASeq/seurat"

#Specify exactly where your matrix files/raw data live
datafolder = "Data"

#Set minimums for genes/cell (min.features) and cell expressing a given gene (min.cells). Default is 1/1, you can filter later. If the object is very large, filter here. 
min.cells = 1
min.features = 1

#Name your project something descriptive
project<-"lesion_snRNASeq" 

sexgene = "Xist"

#Add in metadata. You may want to run the next chunk and then return to edit this so it is in the same order as "ids"
#age = c(rep("P5",2),rep("P7",2))
#condition = c("Deprived","Control","Deprived","Control")

#metrics you want to look at for QC
m = c("nCount_RNA","nFeature_RNA","percent.mito")

#set up folders
QCfolder = "QC"
```

#load 10x data from each *already demultipelxed* sample
```{r}
samples = c("1_Rest_Ctrl", "2_2dpi_Ctrl", "3_7dpi_Ctrl", "4_21dpi_Ctrl", "5_7dpi_cKO", "6_21dpi_cKO")
subsamples = c("Rest_ctrl_meninges", "Rest_ctrl_parenchyma", "2dpi_ctrl_lesion", "2dpi_ctrl_parenchyma", "7dpi_ctrl_lesion", "7dpi_ctrl_parenchyma", "21dpi_ctrl_lesion", "21dpi_ctrl_parenchyma", "7dpi_cKO_lesion", "7dpi_cKO_parenchyma", "21dpi_cKO_lesion", "21dpi_cKO_parenchyma")

#metadata to add
timepoint = c("Rest", "2dpi", "7dpi", "21dpi", "7dpi", "21dpi")
genotype = c("WT", "WT", "WT", "WT", "cKO", "cKO")
organ = c("Meninges", "Cortex","Cortex","Cortex","Cortex","Cortex", "Cortex","Cortex","Cortex","Cortex","Cortex", "Cortex")
microanatomy = c("Meninges", "Parenchyma", "Lesion", "Parenchyma","Lesion", "Parenchyma","Lesion", "Parenchyma","Lesion", "Parenchyma","Lesion", "Parenchyma")

sobject_list = list()

subsample_iterate = 1
for(sample_iterate in 1:length(samples)){
  
  sample = samples[sample_iterate]
  subsample = subsamples[subsample_iterate]
  
  data.dir = file.path("/mnt/DATA/molofsky_lab/newing-crystal/Lesion_snRNASeq/cellranger_runs", sample, "outs/per_sample_outs", subsample, "/count/sample_filtered_feature_bc_matrix")

  data.10x <- Read10X(
    data.dir,
    gene.column = 2,
    cell.column = 1,
    unique.features = TRUE,
    strip.suffix = FALSE
  )
  
  sobject <- CreateSeuratObject(
        data.10x$`Gene Expression`,
        project = project, 
        min.cells = min.cells, 
        min.features = min.features)
  
  #NOTE - using demultiplexed data from cellranger multi, so only including multiplexing capture for reference. NOT using Seurat to perform demultiplexing
  sobject[['CMO']] = CreateAssayObject(counts = data.10x$`Multiplexing Capture`)
  
  #add metadata
  sobject$sample = samples[sample_iterate]
  sobject$subsample = subsamples[subsample_iterate]
  sobject$timepoint = timepoint[sample_iterate]
  sobject$genotype = genotype[sample_iterate]
  sobject$organ = organ[subsample_iterate]
  sobject$microanatomy = microanatomy[subsample_iterate]
  sobject <- PercentageFeatureSet(sobject, pattern = "^mt-", col.name = "percent.mito")
  sobject <- PercentageFeatureSet(sobject, pattern = "^Rp", col.name = "percent.ribo")
  
  sex = GetAssayData(sobject,slot = "counts")[sexgene,]
  sex[sex>0] = "Female"
  sex[sex != "Female"] = "Male"
  sobject$sex = sex

  sobject_list[subsample_iterate] = sobject
  
  subsample_iterate = subsample_iterate + 1
  subsample = subsamples[subsample_iterate]

  data.dir = file.path("/mnt/DATA/molofsky_lab/newing-crystal/Lesion_snRNASeq/cellranger_runs", sample, "outs/per_sample_outs", subsample, "/count/sample_filtered_feature_bc_matrix")

  data.10x <- Read10X(
    data.dir,
    gene.column = 2,
    cell.column = 1,
    unique.features = TRUE,
    strip.suffix = FALSE
  )
  
  sobject <- CreateSeuratObject(
        data.10x$`Gene Expression`,
        project = project, 
        min.cells = min.cells, 
        min.features = min.features)
  
  #NOTE - using demultiplexed data from cellranger multi, so only including multiplexing capture for reference. NOT using Seurat to perform demultiplexing
  sobject[['CMO']] = CreateAssayObject(counts = data.10x$`Multiplexing Capture`)
  
  #add metadata
  sobject$sample = samples[sample_iterate]
  sobject$subsample = subsamples[subsample_iterate]
  sobject$timepoint = timepoint[sample_iterate]
  sobject$genotype = genotype[sample_iterate]
  sobject$organ = organ[subsample_iterate]
  sobject$microanatomy = microanatomy[subsample_iterate]
  sobject <- PercentageFeatureSet(sobject, pattern = "^mt-", col.name = "percent.mito")
  sobject <- PercentageFeatureSet(sobject, pattern = "^Rp", col.name = "percent.ribo")
  
  sex = GetAssayData(sobject,slot = "counts")[sexgene,]
  sex[sex>0] = "Female"
  sex[sex != "Female"] = "Male"
  sobject$sex = sex

  sobject_list[subsample_iterate] = sobject
  
  subsample_iterate = subsample_iterate + 1
}

```
#Add special metadata for cKOs (9-12)
#NOTE - multiple genotypes pooled in single sample, separate by sex
#NOTE - these samples NOT ANALYZED in publication (low yield, imperfect genotype segregation)
```{r}
sobject = sobject_list[[9]]
genotype = sobject$sex
genotype[which(genotype == "Female")] = "Col1a2creER Tgfbr2flox"
genotype[which(genotype == "Male")] = "Cdh5creER Tgfbr2flox"
sobject$genotype = genotype
sobject_list[[9]] = sobject

sobject = sobject_list[[10]]
genotype = sobject$sex
genotype[which(genotype == "Female")] = "Col1a2creER Tgfbr2flox"
genotype[which(genotype == "Male")] = "Cdh5creER Tgfbr2flox"
sobject$genotype = genotype
sobject_list[[10]] = sobject

sobject = sobject_list[[11]]
genotype = sobject$sex
genotype[which(genotype == "Female")] = "Cdh5creER Tgfbr2flox"
genotype[which(genotype == "Male")] = "Cdh5creER Col1a2creER Tgfbr2flox"
sobject$genotype = genotype
sobject_list[[11]] = sobject

sobject = sobject_list[[12]]
genotype = sobject$sex
genotype[which(genotype == "Female")] = "Cdh5creER Tgfbr2flox"
genotype[which(genotype == "Male")] = "Cdh5creER Col1a2creER Tgfbr2flox"
sobject$genotype = genotype
sobject_list[[12]] = sobject
```

#Merge
```{r}
sobject.merge = merge(sobject_list[[1]], y=c(sobject_list[[2]], sobject_list[[3]], sobject_list[[4]], sobject_list[[5]], sobject_list[[6]], sobject_list[[7]], sobject_list[[8]], sobject_list[[9]], sobject_list[[10]], sobject_list[[11]], sobject_list[[12]]), add.cell.ids = subsamples)
```

#QC

1. Function to make an annotated violin plot
```{r}
anno_vln = function(seurat_object = sobject.merge, metric = "nCount_RNA", group = "orig.ident"){
  #specify number of categories
  #n = length(levels(seurat_object@meta.data[,group]))
  #was throwing errors, corrected to 1
  n=1

  #make a violin plot
  a = VlnPlot(seurat_object,metric,group.by = group,pt.size = 0,log = F,cols = rep("lightgrey",n)) + scale_fill_manual(values="lightgrey")

  #make a simple function that quantifies percentiles for each column
  add.percentiles = function(x) {
    r = quantile(x, probs = seq(0,1,0.2))
    r
  }

  a + geom_boxplot(width = 0.05,outlier.size = 0,notchwidth = 10,notch = F) + theme_bw() + stat_summary(fun=add.percentiles, geom="point", pch="_", colour=c(rep(c("red3","purple","yellow","forestgreen","darkblue","grey"),n)), size=10)
}
```

Plot each metric. Red = 0%, Purple = 20%, Yellow = 40%, Black = 50%,Green = 60%, Blue = 80%, Grey = 100% percentile
```{r}
#Save graphs
for (i in m){
  graph = anno_vln(metric = i)
  setEPS()
  postscript(file.path(dir,QCfolder,paste0(i,project,".eps")))
  print(graph)
  dev.off()
}

#print graphs to console
for (i in m){
  graph = anno_vln(metric = i)
  print(graph)
}
```


Plot the number of cells each gene is represented by - i.e. how many cells express any given gene? Ordered from rare to common genes. So you might consider focusing on genes that are relatively well expressed across the sample, or you might care about genes expressed only in a small subset of cells. The >=2 refers to read number per cell. 
```{r}
plot(sort(Matrix::rowSums(GetAssayData(sobject.merge)>=2)),xlab="gene rank",ylab="number of cells",main="Cells per genes ( >= 2 )")
```

A. Gene Plot, scatter plot of gene expression across cells, (colored by sample) - sequencing depth relative to genes; looking for similarity across samples. Some cell types can have relatively few genes, but highly expressed, could show up as a second arc below the one seen here (should contain all samples though)
B. % Mitochondrial RNA vs Feature counts
```{r}
Idents(sobject.merge) = "subsample"
#Save graphs
for (i in 1:(length(m)-1)){
  graph = FeatureScatter(sobject.merge,m[i],m[i+1])+ theme_bw()
  setEPS()
  postscript(file.path(dir,QCfolder,paste0(m[i],"_",m[i+1],project,"scatter.eps")))
  print(graph)
  dev.off()
}

#print graphs
for (i in 1:(length(m)-1)){
  graph = FeatureScatter(sobject.merge,m[i],m[i+1])+ theme_bw()
  print(graph)
}
```
Set limits based on the graphs
```{r}
mlo = 0.0000
mhi = quantile(sobject.merge$percent.mito, 0.975)
clo = quantile(sobject.merge$nCount_RNA, 0.025)
chi = quantile(sobject.merge$nCount_RNA, 0.975)
flo = quantile(sobject.merge$nFeature_RNA, 0.025) 
fhi = quantile(sobject.merge$nFeature_RNA, 0.975) 
```

Violin plot of 1) number of genes, 2) number of UMI and 3) percent mitochondrial genes; save with lines that show cutoffs

```{r}
group = "sample_description"
#Make a table with your cutoff values
tablelow = c(clo,flo,mlo)
tablehigh = c(chi,fhi,mhi)
t = as.data.frame(rbind(tablelow,tablehigh))
colnames(t) = m


#save graphs
for (i in 1:ncol(t)){
  metric = colnames(t)[i]
  low = t[1,i]
  high = t[2,i]
  graph = anno_vln(metric = metric) + geom_abline(intercept = high, slope = 0, linetype = "dashed",color = "red", size = 0.8) + geom_abline(intercept = low, slope = 0, linetype = "dashed",color = "black", size = .8)
  setEPS()
  postscript(file.path(dir,QCfolder,paste0(colnames(t)[i],project,"withcutoffs.eps")))
  print(graph)
  dev.off()
}

#print graphs
for (i in 1:ncol(t)){
   metric = colnames(t)[i]
  low = t[1,i]
  high = t[2,i]
  graph = anno_vln(metric = metric) + geom_abline(intercept = high, slope = 0, linetype = "dashed",color = "red", size = 0.8) + geom_abline(intercept = low, slope = 0, linetype = "dashed",color = "black", size = .8)
  print(graph)
}

```


Scatter plots with lines to show cutoffs
```{r}
#Save graphs
for (i in 1:(length(colnames(t))-1)){
  xm = colnames(t)[i]
  ym = colnames(t)[i+1]
  xlow = t[1,i]
  xhigh = t[2,i]
  ylow = t[1,i+1]
  yhigh = t[2,i+1]
  graph = FeatureScatter(sobject.merge,xm,ym)+ theme_bw() + geom_vline(xintercept = xlow, linetype = "dashed",color = "black", size = .8) + geom_vline(xintercept = xhigh, linetype = "dashed",color = "red", size = .8)+ geom_hline(yintercept = ylow, linetype = "dashed",color = "black", size = .8)  + geom_hline(yintercept = yhigh, linetype = "dashed",color = "red", size = .8)
  setEPS()
  postscript(file.path(dir,QCfolder,paste0(m[i],"_",m[i+1],project,"scatter_withcutoffs.eps")))
  print(graph)
  dev.off()
}

#print graphs
for (i in 1:(length(colnames(t))-1)){
  xm = colnames(t)[i]
  ym = colnames(t)[i+1]
  xlow = t[1,i]
  xhigh = t[2,i]
  
  ylow = t[1,i+1]
  yhigh = t[2,i+1]
  
  graph = FeatureScatter(sobject.merge,xm,ym)+ theme_bw() + geom_vline(xintercept = xlow, linetype = "dashed",color = "black", size = .8) + geom_vline(xintercept = xhigh, linetype = "dashed",color = "red", size = .8)+ geom_hline(yintercept = ylow, linetype = "dashed",color = "black", size = .8)  + geom_hline(yintercept = yhigh, linetype = "dashed",color = "red", size = .8)
  print(graph)
}
```


Filtering cells: note that they are taking these filters and duplicating them across all samples. Use -Inf instead of 0 because not all the filters have to be positive values, and to avoid any errors due to truncation. 

How to select filters??? noise vs signal - get them from the violin plots and the matrix, capture the bulk of cells but exclude outliers (say with mtDNA filter) - can also look at percentiles, maybe get rid of top 5% since it was very different from the rest. Basically eliminate outliers and don't include cells that have hardly any genes or have a ton of genes expressed. First save the original object: 

```{r}
save(sobject,file = file.path(dir,"Data/Seurat",paste0(project,"_beforeQC.RData")))
```

Subset the data. Printing tables lets you see the progression as you tighten your QC cutoffs. 
```{r}
print("initial")
table(sobject.merge$subsample) #print out initial cell counts
sobject.merge = subset(sobject.merge, subset = percent.mito>mlo & percent.mito < mhi) #subset mitochondria
print("mito")
table(sobject.merge$subsample) #new cell counts
sobject.merge = subset(sobject.merge, subset = nCount_RNA>clo & nCount_RNA < chi)
print("ncount")
table(sobject.merge$subsample)
sobject.merge = subset(sobject.merge, subset = nFeature_RNA>flo & nFeature_RNA < chi)
print("nfeature")
table(sobject.merge$subsample)
```
```{r}
save(sobject.merge, file = file.path(dir,"Data/Seurat",paste0(project,"_filtered.RData")))
```