---
title: "15_CellPhone_DB"
output: html_document
date: "2023-06-05"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(Seurat)

require(tibble)
require(biomaRt)
require(tidyr)
require(dplyr)
library(data.table)
```

```{r}
#Specify your preferred directory for all input + output
dir = "/Users/nathanewing-crystal/Dropbox (Ari Molofsky Lab)/Ari Molofsky Lab Team Folder/Nathan Ewing-Crystal/RNASeq/Lesion snRNASeq/seurat"

#Specify exactly where your matrix files/raw data live
datafolder = "Data/Seurat"
Plotfolder = "Plots/CellPhoneDB"
project<-"lesion_snRNASeq"
#load object: use #12 final_fib-immune-astro-subset
iterationname = "final_fib-immune-astro-subset"
base_name = paste0(project, iterationname, "_cellphonedb_")
load(file.choose())

#for WT:
Idents(sobject) = "genotype"
sobject.wt = subset(sobject, ident = "WT")
```

```{r}
name = paste0(project, iterationname)
DefaultAssay(sobject.wt) = "RNA"
Idents(sobject.wt) = "celltype_subset"
```

```{r}
# Basic function to convert mouse to human gene names
alldata <- NormalizeData(sobject.wt, normalization.method = "LogNormalize", scale.factor = 10000)

matrix1 <- as.data.frame(alldata@assays$RNA@data)
matrix1 <- matrix1[rowSums(matrix1[,2:dim(matrix1)[2]])!=0,]

### If you are using a mouse data, then its needed to convert the gene names to human orthologs

human = useMart("ensembl", dataset = "hsapiens_gene_ensembl", host = "dec2021.archive.ensembl.org")
mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl", host = "dec2021.archive.ensembl.org")

#change to rownames(matrix1) from rownames(alldata@assays$RNA@data)
genesV2 = getLDS(attributes = c("mgi_symbol"), filters = "mgi_symbol", values = rownames(matrix1) , mart = mouse, attributesL = c("hgnc_symbol","hgnc_id",'ensembl_gene_id'), martL = human, uniqueRows=T)
print(head(genesV2))

index = match(genesV2$MGI.symbol,rownames(matrix1),nomatch=F)
matrix2 <- matrix1[index,]
#genesV2 includes mystery gene at index 14959 (Etv3l) that is not in allgenes or rownames(matrix1); remove
genesV2 <- genesV2[-14959,]

#duplicate human values; to solve, order by "max expression" and take unique values (with duplicates, will take highest expressed only)
matrix3 = data.frame(rownames(matrix2), genesV2$HGNC.symbol, genesV2$MGI.symbol)
matrix3$max = apply(matrix2, 1, max, na.rm=TRUE)
matrix4 = matrix3[order(-matrix3$max),]
matrix5 = matrix4[!duplicated(matrix4$genesV2.HGNC.symbol),]
matrix6 = matrix5[!(matrix5$genesV2.HGNC.symbol == ""), ] #remove empty

#check
table(matrix6$genesV2.MGI.symbol == matrix6$rownames.matrix2.)
  #some duplicates, i.e. 
# > matrix5[which(matrix5$rownames.matrix2. == "7420426K07Rik.2"),]
#       rownames.matrix2. genesV2.HGNC.symbol genesV2.MGI.symbol       max
# 16534   7420426K07Rik.2              PRR23C      7420426K07Rik 0.2739728
# > matrix5[which(matrix5$rownames.matrix2. == "7420426K07Rik.1"),]
#       rownames.matrix2. genesV2.HGNC.symbol genesV2.MGI.symbol       max
# 16533   7420426K07Rik.1              PRR23B      7420426K07Rik 0.2739728

matrix7 = matrix2[matrix6$rownames.matrix2.,] #retrieve counts (from matrix 2) corresponding to sorted, unique unique genes (stored in matrix 5)

rownames(matrix7) = matrix6$genesV2.HGNC.symbol #will be in correct order because ordered matrix6 by matrix5 order

setDT(matrix7, keep.rownames = "Gene")[]

write.table(matrix7, 
            file = '/Users/nathanewing-crystal/Dropbox (Ari Molofsky Lab)/Ari Molofsky Lab Team Folder/Nathan Ewing-Crystal/RNASeq/Lesion snRNASeq/seurat/Data/cellphonedb/lesion_snRNASeqfinal_fib-immune-astro-subset_counts_updated_WT_ONLY.txt',
            quote=F,
            col.names = T,
            row.names=F,
            sep='\t')

metadata = as.data.frame(alldata$celltype_subset)
setDT(metadata, keep.rownames = "Cell")[]
colnames(metadata)[2] = "cell_type"

write.table(metadata, 
            file = '/Users/nathanewing-crystal/Dropbox (Ari Molofsky Lab)/Ari Molofsky Lab Team Folder/Nathan Ewing-Crystal/RNASeq/Lesion snRNASeq/seurat/Data/cellphonedb/lesion_snRNASeqfinal_fib-immune-astro-subset_metadata_WT_ONLY.txt',
            quote=F,
            col.names = T,
            row.names=F,
            sep='\t')
```

#actual cellphoneDB analysis in Jupityr notebook
# http://localhost:8888/notebooks/cellphone_db.ipynb

```{r}
#if (!requireNamespace("devtools", quietly = TRUE))
#    install.packages("devtools")
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#devtools::install_github('zktuong/ktplots', dependencies = TRUE)

library(ktplots)
```

```{r}
#Note - using "check.names=FALSE to avoid converting "-" to "." in cluster titles

means = read.delim("/Users/nathanewing-crystal/Dropbox (Ari Molofsky Lab)/Ari Molofsky Lab Team Folder/Nathan Ewing-Crystal/RNASeq/Lesion snRNASeq/seurat/Data/cellphonedb/Results/statistical_analysis_means_06_28_2023_11:53:31.txt", sep="\t", check.names=FALSE)
pvals = read.delim("/Users/nathanewing-crystal/Dropbox (Ari Molofsky Lab)/Ari Molofsky Lab Team Folder/Nathan Ewing-Crystal/RNASeq/Lesion snRNASeq/seurat/Data/cellphonedb/Results/statistical_analysis_pvalues_06_28_2023_11:53:31.txt", sep="\t", check.names=FALSE)
decon = read.delim("/Users/nathanewing-crystal/Dropbox (Ari Molofsky Lab)/Ari Molofsky Lab Team Folder/Nathan Ewing-Crystal/RNASeq/Lesion snRNASeq/seurat/Data/cellphonedb/Results/statistical_analysis_deconvoluted_06_28_2023_11:53:31.txt", sep="\t", check.names=FALSE)

#read in python-generated file containing all significant interactions
all_interactions = read.csv("/Users/nathanewing-crystal/Dropbox (Ari Molofsky Lab)/Ari Molofsky Lab Team Folder/Nathan Ewing-Crystal/RNASeq/Lesion snRNASeq/seurat/Data/cellphonedb/Results/all_interactions.csv")
```

#start with Fibroblast_Cd80
```{r}
PlotfolderOld = "Plots/CellPhoneDB/Manual"

Idents(sobject.wt) = "celltype_subset"

families = c('chemokines', 'Th1', 'Th2', 'Th17', 'Treg', 'costimulatory', 'coinhibitory', 'niche')
comparisons = data.frame(matrix(nrow = 5, ncol=0))
rownames(comparisons) = c("cell_type_1", "cell_type_2", "genes", "genes_name", "gene.family") #genes_name to name plots with multiple genes (gestalt)

#first, predefined gene families
for(family in families){
  index = which(families == family)
  colname = paste0("comparison", index)
  comparison = list('Fibroblast_Cd80', 'T-cells_Th1-2', "NULL", "NULL", family)
  comparisons = cbind(comparisons, as.matrix(comparison))
  colnames(comparisons)[index] = colname
}

#add other gene comparisons
genes = c("CCL19", "CX3CL1", "CCL8", "CXCL12", "CXCL16", "IL7", "TSLP", "IGF1", "IGF1", "IL33")
comparison_gene = list('Fibroblast_Cd80', 'T-cells_Th1-2', genes, "candidates", "NULL")
comparisons = cbind(comparisons, "comparison_gene" = as.matrix(comparison_gene))

#for all comparisons, make plots
for(comparison in colnames(comparisons)){
  #set up variables to plot
  cell_type_1 = comparisons[[comparison]][[1]]
  cell_type_2 = comparisons[[comparison]][[2]]
  genes = comparisons[[comparison]][[3]]
  genes_name = comparisons[[comparison]][[4]]
  gene.family = comparisons[[comparison]][[5]]
  
  folder_comparison = paste0(cell_type_1, "_and_", cell_type_2)
  newPlotfolder = file.path(PlotfolderOld, folder_comparison)
  Plotfolder_total = file.path(dir, newPlotfolder)
  if (!dir.exists(Plotfolder_total)){
    print(paste0("Making new Plotfolder: ", newPlotfolder))
    dir.create(Plotfolder_total, recursive=TRUE)
  }
  Plotfolder = newPlotfolder
  
  if("NULL" %in% genes){
    name = paste0(base_name, cell_type_1, "_", cell_type_2, "_test-", gene.family, ".eps")
    p1 = plot_cpdb(cell_type1 = cell_type_1, cell_type2 = cell_type_2,
          scdata = "sobject.wt",
          idents = 'cell_type', # column name where the cell ids are located in the metadata
          means = means, pvals = pvals, max_size = 5,
          gene.family = gene.family,
          scale=FALSE)
  } else{
    name = paste0(base_name, cell_type_1, "_", cell_type_2, "_test-", genes_name, ".eps")
    p1 = plot_cpdb(cell_type1 = cell_type_1, cell_type2 = cell_type_2,
          scdata = "sobject.wt",
          idents = 'cell_type', # column name where the cell ids are located in the metadata
          means = means, pvals = pvals, max_size = 5,
          genes = genes, 
          scale=FALSE)
  }
  
  extent = length(rownames(p1$data))/2
  if(extent >20){ #replot with smaller
    max_size = 100/extent
    fontsize = 300/extent
    if("NULL" %in% genes){
      name = paste0(base_name, cell_type_1, "_", cell_type_2, "_test-", gene.family, ".eps")
      p1 = plot_cpdb(cell_type1 = cell_type_1, cell_type2 = cell_type_2,
            scdata = "sobject.wt",
            idents = 'cell_type', # column name where the cell ids are located in the metadata
            means = means, pvals = pvals, max_size = max_size,
            gene.family = gene.family,
            scale=FALSE) + small_axis(fontsize = fontsize)
    } else{
      name = paste0(base_name, cell_type_1, "_", cell_type_2, "_test-", genes_name, ".eps")
      p1 = plot_cpdb(cell_type1 = cell_type_1, cell_type2 = cell_type_2,
            scdata = "sobject.wt",
            idents = 'cell_type', # column name where the cell ids are located in the metadata
            means = means, pvals = pvals, max_size = max_size,
            genes = genes,
            scale=FALSE) + small_axis(fontsize = fontsize)
    }
  }
  
  setEPS()
  postscript(file.path(dir, Plotfolder, name))
  print(p1)
  dev.off()
}


```

```{r}
plot_cpdb_total("Fibroblast_Cd80|Fibroblast_Fn1", "T-cells_Th1-2|T-cells_Th17", c("HLA E", "VCAM1", "CD200", "IL18", "ICAM1", "CXCL12", "CD80", "SPP1", "IL7", "IL15", "CXCL9", "CXCL10", "CCL8", "CXCL16"))

plot_cpdb_total("Fibroblast_Cd80", "T-cells_Th1-2|T-cells_Th17", c("HLA E", "VCAM1", "CD200", "IL18", "ICAM1", "CXCL12", "CD80", "SPP1", "IL7", "IL15", "CXCL9", "CXCL10", "CCL8", "CXCL16"))

plot_cpdb_total("Macrophage_Ldlrad3|DAMs", "Fibroblast_Fn1", c("TGFB1", "TGFB2", "TGFB3", "PDGFA", "PDGFB", "PDGFC", "CXCL12", "ICAM1", "VCAM1", "CSF1", "IGF1", "SPP1", "FN1", "FBN1", "APOE", "SIRPA", "IL1A", "IL1B"))

plot_cpdb_total("Monocytes_DCs", "Fibroblast_Fn1", c("TGFB1", "TGFB2", "TGFB3", "PDGFA", "PDGFB", "PDGFC", "CXCL12", "ICAM1", "VCAM1", "CSF1", "IGF1", "SPP1", "FN1", "FBN1", "APOE", "SIRPA", "IL1A", "IL1B"))

plot_cpdb_total("Astrocyte_Cd84", "T-cells_Th1-2|T-cells_Th17", c("IL18", "CXCL16", "LTB", "TNF", "CD80", "CXCL12", "CD86", "HLA #", "ICAM1", "SPP1"))

plot_cpdb_total("Fibroblast_Fn1", "Fibroblast_Fn1", c("COL1A1", "COL1A2", "FN1", "TGFB1", "TGFB2", "TGFB3", "IGF1", "SPP1", "TNC"))
```


#get genes involved in significant interactions
```{r}
get_genes = function(celltype1, celltype2){
  
  if(celltype1 == "all"){celltype1 = "*"}
  if(celltype2 == "all"){celltype2 = "*"}  
  
  interactions = all_interactions[intersect(grep(celltype1, all_interactions$interacting_cells), grep(celltype2, all_interactions$interacting_cells)),]
  
  genes = union(interactions$gene_a, interactions$gene_b)
  genes = genes[genes != ""]
  return(genes)
}
```

#plot significant (or otherwise specified) interactions
```{r}
plot_interactions = function(cell_type_1, cell_type_2, genes_name = NULL, significant = TRUE, genes = NULL, seurat_object = sobject.wt){
  
  if(significant == TRUE){
    if(is.null(genes)){
      genes_name = "significant"
      genes = get_genes(cell_type_1, cell_type_2)
    } else{
      if(is.null(genes_name)){
        print("No genes_name specified")
        return()
      } else{
        genes_name = paste0(genes_name, "significant")
      }
    }
    keep_significant_only	= TRUE
  } else {
    if(is.null(genes)){
      print("No genes specified")
      return()
    }
    if(is.null(genes_name)){
      print("No genes_name specified")
      return()
    }
    keep_significant_only	= FALSE
  }
  
  #set up directory
  PlotfolderOld = "Plots/CellPhoneDB"
  folder_comparison = paste0(cell_type_1, "_and_", cell_type_2)
  newPlotfolder = file.path(PlotfolderOld, folder_comparison)
  Plotfolder_total = file.path(dir, newPlotfolder)
  if (!dir.exists(Plotfolder_total)){
     print(paste0("Making new Plotfolder: ", newPlotfolder))
     dir.create(Plotfolder_total, recursive=TRUE)
  }
  Plotfolder = newPlotfolder
    
  name = paste0(base_name, cell_type_1, "_", cell_type_2, "_test-", genes_name, ".eps")

  p1 = plot_cpdb(cell_type1 = cell_type_1, cell_type2 = cell_type_2,
            scdata = "sobject.wt",
            idents = 'cell_type', # column name where the cell ids are located in the metadata
            means = means, pvals = pvals, max_size = 5,
            genes = genes, 
            scale=FALSE,
            highlight_size = 0.75,
            keep_significant_only	= keep_significant_only)
    
  extent = (if(significant==TRUE){length(is.na(p1$data$pvals))} else{length(rownames(p1$data))})
  if(extent >30){ #replot with smaller
    max_size = min(6, 1200/extent)
    fontsize = min(6, 2000/extent)
    
    p1 = plot_cpdb(cell_type1 = cell_type_1, cell_type2 = cell_type_2,
              scdata = "sobject.wt",
              idents = 'cell_type', # column name where the cell ids are located in the metadata
              means = means, pvals = pvals, max_size = max_size,
              genes = genes,
              scale=FALSE,
              highlight_size = 0.75*(max_size/5),
              keep_significant_only	= keep_significant_only) + small_axis(fontsize = fontsize)
    }
    
    setEPS()
    postscript(file.path(dir, Plotfolder, name))
    print(p1)
    dev.off()
}
```

```{r}
plot_cpdb_total = function(celltype1, celltype2, genes){
  if(is.null(genes)){
    print("Missing \"genes\" argument")
    return
  }
  plot_interactions(celltype1, celltype2)
  plot_interactions(celltype1, celltype2, significant=F, genes_name = "specific", genes = genes)
  plot_interactions(celltype1, celltype2, significant=T, genes_name = "specific", genes = genes)
}
```

```{r}
#Figure
Plotfolder = "Plots/figure"

genes= c("TGFB1", "TGFB2", "TGFB3", "PDGFA", "PDGFB", "PDGFC", "CSF1", "IGF1", "FN1", "FBN1", "APOE", "SIRPA")

p1 = plot_cpdb(cell_type1 = "Macrophage_Ldlrad3|DAMs", cell_type2 = "Fibroblast_Fn1",
          scdata = "sobject.wt",
          idents = 'cell_type', # column name where the cell ids are located in the metadata
          means = means, pvals = pvals, max_size = 5,
          genes = genes, 
          scale=T,
          keep_significant_only = T,
          exclude_interactions = NULL)

#exclude
interactions = c(
  "FN1-integrin-a11b1-complex", 
  "IGF2-IGF1R", 
  "FBN1-integrin-a5b1-complex", 
  "FN1-integrin-a4b1-complex",  
  "FN1-integrin-aVb1-complex",  
  "FN1-integrin-aVb5-complex",
  "TGFB3-integrin-aVb8-complex",
  "TGFB3-TGFBR3", "TGFB2-TGFBR3", "TGFB1-TGFBR3",
  "TGFB3-integrin-aVb6-complex", "TGFB1-integrin-aVb6-complex",
  "PDGFB-PDGFRB", "PDGFB-PDGFR-complex",
  "IGF1-integrin-a6b4-complex",
  "IGF1-IGF1R-enhancerComGPC3",
  "FN1-integrin-a8b1-complex", "FN1-integrin-a4b7-complex", "IGF1-integrin-aVb3-complex", "FN1-integrin-aVb3-complex", "FN1-integrin-a3b1-complex", "IL34-CSF1R",
  "FN1-integrin-a2Bb3-complex",
  "FN1-integrin-a2Bb3-complex",
  "FN1-integrin-a2b1-complex")

p2 = plot_cpdb(cell_type1 = "Macrophage_Ldlrad3|DAMs", cell_type2 = "Fibroblast_Fn1",
          scdata = "sobject.wt",
          idents = 'cell_type', # column name where the cell ids are located in the metadata
          means = means, pvals = pvals, max_size = 5,
          genes = genes, 
          scale=T,
          keep_significant_only = F,
          exclude_interactions = interactions,
          cluster_rows = F,
          highlight_size = 1)

p2$data$Var2 = factor(p2$data$Var2, levels = c("DAMs-Fibroblast_Fn1", "Macrophage_Ldlrad3-Fibroblast_Fn1", "Fibroblast_Fn1-DAMs", "Fibroblast_Fn1-Macrophage_Ldlrad3"))

levels = levels(p2$data$Var1)
levels_keep = c(
  "APOE-TREM2-receptor",        

  "CSF1-CSF1R",                 

  "SIRPA-CD47",
  
  "IGF1-IGF1R",                 
  
  "PDGFC-PDGFRA", 
  "PDGFB-PDGFRA",               
  "PDGFA-PDGFRA",  
  
  "TGFB3-TGFbeta-receptor2",
  "TGFB3-TGFbeta-receptor1", 
  "TGFB2-TGFbeta-receptor2",
  "TGFB2-TGFbeta-receptor1",
  "TGFB1-TGFBR3",  
  "TGFB1-TGFbeta-receptor2",
  "TGFB1-TGFbeta-receptor1"    
)
levels_leave = levels[!(levels %in% levels_keep)]
levels_replace = c(levels_leave, levels_keep)

p2$data$Var1 = factor(p2$data$Var1, levels = levels_replace)

setEPS()
postscript(file.path(dir, Plotfolder, "cpdb_tgfb"), width = 5, height = 4.5)
p2 
dev.off()

setEPS()
postscript(file.path(dir, Plotfolder, "cpdb_tgfb_LEGEND"), width = 4.25, height = 12)
p2 
dev.off()

```

