---
title: "Fib subset"
author: "Nathan Ewing-Crystal"
date: "2023-05-15"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
#if (!require("BiocManager", quietly = TRUE))
    #install.packages("BiocManager")

#BiocManager::install("variancePartition")

library(Seurat)
library(ggplot2)
library(dplyr)
library(ape)
library(cowplot)
library(Matrix)
library(variancePartition)
library(MAST)
#install.packages('EnhancedVolcano')
library(EnhancedVolcano)

#BiocManager::install("goseq")
library(goseq)
#BiocManager::install("TxDb.Mmusculus.UCSC.mm10.ensGene")
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
#BiocManager::install("org.Mm.eg.db")
library(org.Mm.eg.db)
#BiocManager::install("clusterProfiler")
library(clusterProfiler)

#BiocManager::install("Nebulosa")
library("Nebulosa")
```

#Load filtered seurat object
```{r}
load(file.choose())
```

```{r}
#Specify your preferred directory for all input + output
dir = "/Users/nathanewing-crystal/Ari Molofsky Lab Dropbox/Ari Molofsky Lab Team Folder/Nathan Ewing-Crystal/RNASeq/Lesion snRNASeq/seurat"
#Specify exactly where your seurat files live
datafolder = "Data/Seurat/Fibroblast Subset"

project<-"lesion_snRNASeq" 

#set up folders
QCfolder = "QC"
Plotfolder = "Plots/Fibroblast Subset"

#Decide whether to use SCtransform (can take a long time with big datasets; generally recommended)
sct = TRUE

#You can add in housekeeping genes if you want them to be scaled always; otherwise set to "NULL"
add.genes = NULL

#How many genes do you want scaled to use for UMAP/clustering? 
ngenes = 6000

#Which principal components do you want to calculate on? This is a default setting, change if one of the pc's is for something you think is a technical error (i.e. HSP, RP, etc)
pcs = c(1:30)

#Choose what to regress out - could be age, sex, or any metadata column
regress = c("nCount_RNA","percent.mito")

#clustering resolution; the last number will be saved as "seurat_clusters" in metadata
res = c(1.5,1.0,0.5, 0.25)

#metadata dimensions you want to cluster on
dims = c("seurat_clusters","sample", "subsample", "timepoint","genotype", "microanatomy")

#make a unique name; maybe this is the celltype you've already subset, or the age you're looking at, etc.
iterationname = "fib_reclustered"

#Establish cutoffs for heatmaps
pval = 1e-3 #max p-value for significance
lfc = 0.2 #minimum log fold change
minpct = 0.05 #if you want to subset by percent cells in that cluster expressing the gene
maxpct = 1
single = F #should each gene be a marker of a single cluster only
hgenes = 3 #how many genes should be in the heatmap per cluster
ncells = 100 #max # of cells per heatmap column
column = "seurat_clusters" #division you care about
```

Find variable features, normalize, scale, run PCA, clustering, umap
The following is the standard method of normalization and scaling. Interchangeable with the next chunk. Run both, you should have already specified which to use with "sct = T or F". Will take 5-20 minutes to run. 
```{r,results = 'hide',eval = FALSE}
if (!sct){
  sobject.sub.fib <- NormalizeData(sobject.sub.fib,normalization.method = "LogNormalize", scale.factor = 10000)
  sobject.sub.fib<-FindVariableFeatures(sobject.sub.fib, selection.method = "vst", nfeatures = ngenes)
  
  all.genes<-rownames(sobject.sub.fib)
  var.genes = VariableFeatures(sobject.sub.fib)
  add.genes = add.genes[!add.genes %in% var.genes]
  
  any(add.genes %in% var.genes)
  scalegenes = c(var.genes,add.genes)
  VariableFeatures(sobject.sub.fib) = scalegenes
  sobject.sub.fib<-ScaleData(sobject.sub.fib,features = VariableFeatures(sobject.sub.fib), vars.to.regress = regress)
}
```

Alternative: SCTransform (great for smaller datasets)
```{r, results='hide',eval = FALSE}
if (sct){
  sobject.sub.fib <- SCTransform(sobject.sub.fib, vars.to.regress = regress, verbose = FALSE,variable.features.n = ngenes,conserve.memory = T)
}
```

Show most variable genes
```{r}
labels <- c(head(VariableFeatures(sobject.sub.fib),10),add.genes)
plot1 = VariableFeaturePlot(sobject.sub.fib)
LabelPoints(plot=plot1, points = labels, repel = F, xnudge = 0.1, ynudge = 0.5)
```

Run PCA analysis and show elbow plot
```{r}
sobject.sub.fib <- RunPCA(sobject.sub.fib,features = VariableFeatures(sobject.sub.fib),npcs = 50, verbose = FALSE)
ElbowPlot(sobject.sub.fib,ndims = 50, reduction = "pca")
print(sobject.sub.fib[["pca"]], dims = 1:20, nfeatures = 5)
```

Once you are satisfied with pc's, run clustering: 
```{r}
sobject.sub.fib<-RunUMAP(sobject.sub.fib,reduction = "pca",dims = pcs, verbose = F)
sobject.sub.fib<-FindNeighbors(sobject.sub.fib,dims=pcs,verbose=F)
sobject.sub.fib<-FindClusters(sobject.sub.fib,verbose=F,resolution = res)
sobject.sub.fib$seurat_clusters = sobject.sub.fib$SCT_snn_res.0.5

```

Function to print multiple graphs: 
```{r}
PrintSeuratGraph = function(namecard = "a",seurat_object = sobject.sub.fib,graphtype = "feature",feature = NULL,group = NULL,split=NULL,cellnames=NULL, label=FALSE){
  if (!is.null(cellnames)){
    #Idents(seurat_object) = cellnames[1]
    #cells = colnames(seurat_object)[Idents(seurat_object) %in% cellnames[1:length(cellnames)]]} 
    cells = cellnames}
  else {cells = cellnames}
  if (graphtype == "feature"){
    graph = FeaturePlot(seurat_object,features = feature,split.by = split, cells = cells,cols = c("lightyellow","darkred"))
  }
  if (graphtype == "violin"){
    graph = VlnPlot(seurat_object,features = feature, pt.size = 0.1, idents = cellnames[2:length(cellnames)],group.by = group, split.by = split)
  }
  if (graphtype == "dim"){
    graph = DimPlot(seurat_object,cells = cells, group.by = group, split.by = split, label = label)
  }
  name = paste0(feature,"_",graphtype,namecard,".eps")
  graph
  setEPS()
  postscript(file.path(dir,Plotfolder,name))
  print(graph)
  dev.off()
}
```

Save the clustered dataset. Overwrite the existing, subset dataset. 
```{r,eval = FALSE}
save(sobject.sub.fib,file = file.path(dir,datafolder,paste0(project,"_",iterationname,".RData")))
```

Block to print multiple graphs: 
NOTE DOUBLETS IN SOME (whichever use "sobj" and not "sobjv2")
```{r}
Plotfolder = "Plots/Fibroblast Subset"
name = paste0(project,"_",iterationname)
genes = c("Col1a2", "Cthrc1", "Lrrc15", "Acta2", "Tgfbi", "Cxcl12", "Il1r1")
groups = c("timepoint_genotype", "timepoint", "microanatomy", "fibtype", "fibtypecluster")
#note - added "fibtype" and "fibtypecluster" from below
genes = genes[genes %in% rownames(GetAssayData(sobject,slot = "data"))]

for(feature in genes){
  PrintSeuratGraph(namecard = name,graphtype = "feature",feature = feature)
}

#dim plots for clustering
for(group in groups){
  PrintSeuratGraph(namecard = name,graphtype = "dim",group = group, feature = group)
}

PrintSeuratGraph(namecard = name, graphtype = "dim", group = "seurat_clusters", feature = "seurat_clusters", label=T)
PrintSeuratGraph(namecard = name, graphtype = "dim", group = "fibtype", feature = "fibtype", label=T)

#violin plots
for(feature in c(genes)){
  PrintSeuratGraph(namecard = name,graphtype = "violin",feature = feature,group = "seurat_clusters")
}

#plot individual timepoint_genotype dim plots, no legends
Plotfolder = "Plots/Fibroblast Subset/Timepoint Dimplots"
for(condition in levels(as.factor(sobject.sub.fib$timepoint_genotype))){
  cells = names(which(sobject.sub.fib$timepoint_genotype == condition))
  graph = DimPlot(sobject.sub.fib, cells = cells, group.by = "seurat_clusters", label = T) + ggplot2::theme(legend.position = "none") + ggtitle(condition) + xlim(-11,13) + ylim(-14, 13)
  name = paste0(condition, "_dimplot.eps")
  setEPS()
  postscript(file.path(dir,Plotfolder,name))
  print(graph)
  dev.off()
}
```

```{r}
#Determine cluster identities

DefaultAssay(sobject.sub.fib) = "RNA"
Plotfolder = "Plots/Fibroblast Subset/Additional Specific Genes"
name = paste0(project,"_",iterationname)

features = c("Mki67", "Ptgds", "Tmeff2", "Fn1", "Nrxn1", "Lama1", "Cd80", "Ghr", "Aldh1a2", "Crabp2", "Ogn", "Col18a1", "Lama1", "Col25a1")
for(feature in features){
   PrintSeuratGraph(namecard = name,graphtype = "feature",feature = feature)
   PrintSeuratGraph(namecard = name, graphtype = "violin", feature = feature)
}

#Cluster 6 - doublets?
sub = subset(sobject.sub.fib, cells = which(sobject.sub.fib$seurat_clusters == "6"))
x = sub@assays$RNA@scale.data["Ptprc",]
y = sub@assays$RNA@scale.data["Col1a2",]
df = data.frame(x,y)

p1 = ggscatter(df, x = "x", y = "y", 
          title = "CD45 vs. Col1a2",
          add = "reg.line", conf.int = TRUE, add.params=list(color="red"),
          cor.coef = TRUE, cor.method = "pearson", cor.coef.size = 5,
          xlab = "CD45", ylab = "Col1a2")
```

```{r}
#Edit this part carefully. You can add any number of types. Each cluster can only be one type. 

type0 = c(-1)
name0 = "Meninges_1"
type1 = c(0)
name1 = "Lama1"
type2 = c(1)
name2 = "Tmeff2"
type3 = c(2)
name3 = "Meninges_2"
type4 = c(3)
name4 = "Ptgds"
type5 = c(4)
name5 = "Fn1"
type6 = c(5)
name6 = "Cd80"
type7 = c(6)
name7 = "Myeloid"
type8 = c(7)
name8 = "Meninges_3"
type9 = c(8)
name9 = "Ghr"
type10 = c(9)
name10 = "Nrxn1-3"
type11 = c(10)
name11 = "Mural_Pericyte"
type12 = c(11)
name12 = "Meninges_4"
type13 = c(12)
name13 = "Prolif"
type14 = c(13)
name14 = "Meninges_5"

#OLD res=0.25
# type0 = c(-1)
# name0 = "Meninges_1"
# type1 = c(0)
# name1 = "Stk39"
# type2 = c(1)
# name2 = "Fn1"
# type3 = c(2)
# name3 = "Adamts12"
# type4 = c(3)
# name4 = "Meninges_2"
# type5 = c(4)
# name5 = "Ptgds"
# type6 = c(5)
# name6 = "Myeloid"
# type7 = c(6)
# name7 = "Meninges_3"
# type8 = c(7)
# name8 = "Meninges_4"
# type9 = c(8)
# name9 = "Nrxn1-3"
# type10 = c(9)
# name10 = "Mural_Pericytes"
# type11 = c(10)
# name11 = "Prolif"

#Initialize the cluster levels as a vector and replace the cluster levels with the appropriate name. 
clusters = as.factor(sobject.sub.fib$seurat_clusters)
type = levels(clusters)
type[type0+2] = name0
type[type1+2] = name1
type[type2+2] = name2
type[type3+2] = name3
type[type4+2] = name4
type[type5+2] = name5
type[type6+2] = name6
type[type7+2] = name7
type[type8+2] = name8
type[type9+2] = name9
type[type10+2] = name10
type[type11+2] = name11
type[type12+2] = name12
type[type13+2] = name13
type[type14+2] = name14

levels(clusters) = type

#Add a metadata column
sobject.sub.fib$fibtype = clusters

#check the celltype assignment for accuracy
table(sobject.sub.fib$fibtype,sobject.sub.fib$seurat_clusters)

#add a metadata column labelling each cluster
sobject.sub.fib$fibtypecluster = paste0(sobject.sub.fib$fibtype,"-",sobject.sub.fib$seurat_clusters)
Idents(sobject.sub.fib) = "fibtype"
```
Save the clustered dataset with fibtype labels. Overwrite the existing, clustered dataset. 
```{r,eval = FALSE}
save(sobject.sub.fib,file = file.path(dir,datafolder,paste0(project,"_",iterationname,".RData")))
```

Save the tree
```{r}
# Idents(sobject.sub.fib) = column
Idents(sobject.sub.fib) = "SCT_snn_res.0.5"
sobject.sub.fib= BuildClusterTree(sobject.sub.fib,dims = pcs)
tree = sobject.sub.fib@tools$BuildClusterTree
setEPS()
postscript(file.path(dir,"Trees",paste0(project,"_",iterationname,"_tree.eps")))
plot.phylo(tree, use.edge.length = T, direction = "rightwards")
dev.off()
```

*Note: already scaled/normalized (before subsetting, in 2_Processing.Rmd)*

*Heatmap*

First calculate DE genes for every cluster (with meninges)
```{r,results = "hide",eval = FALSE}
Idents(sobject.sub.fib) = column
markers_all <- FindAllMarkers(
    object = sobject.sub.fib,
    assay = "RNA",  ##Edited to do DEG analysis on RNA counts, not SCT transformed data
    features = rownames(sobject),
    test.use = "MAST",
    only.pos = FALSE, 
    min.pct = 0.05, 
    logfc.threshold = 0.2
)
write.csv(markers_all,file = file.path(dir,"Spreadsheets",paste0(project, "_", iterationname, "_all_markers.csv")))
```

***~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~***
Before further analysis:
1) exclude cluster 7 (mac doublets); also exclude cluster 10 (neuron doublets)

then:

2) make WT-all subset for pictures (including meningeal cells)

-exclude meningeal clusters (not useful for timecourse, would be useful for cKO vs. WT except didn't include meninges in 7dpi samples)
-exclude mural cells

3) make WT subset to compare timepoints
4) make 7dpi subset to compare genotypes
***~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~***
```{r}
#Exclude unwanted clusters
Idents(sobject.sub.fib) = "seurat_clusters"

clusters_minus_710 = c(0:6, 8:9, 11:14)
sobject.sub.fib.v2 = subset(sobject.sub.fib, idents = clusters_minus_710)
#1)
save(sobject.sub.fib.v2,file = file.path(dir,datafolder,paste0(project,"_",iterationname,"_v2.RData")))

Idents(sobject.sub.fib.v2) = "genotype"
sobject.sub.fib.wt.all = subset(sobject.sub.fib.v2, idents = "WT")
#2)
save(sobject.sub.fib.wt.all,file = file.path(dir,datafolder,paste0(project,"_",iterationname,"_WT-all.RData")))

#Note - excluding meninges by cluster AND microanatomy - some "meningeal" fibroblasts in lesion (likely stuck to lesion during dissection), not useful for comparison between timepoints
clusters_to_keep = c(1:2, 4:6, 9, 11, 13) #exclude clusters 0, 3, 8, 12, 14 (meninges), 7 (myeloid doublets), and 10 (neuronal doublets) = exclude 0, 3, 7, 8, 10, 12, 14
#Note - not saving below "v3" sobject, but using to subset
sobject.sub.fib.v3 = subset(sobject.sub.fib, idents = clusters_to_keep)

Idents(sobject.sub.fib.v3) = "microanatomy"
sobject.sub.fib.v4 = subset(sobject.sub.fib.v3, idents = c("Lesion", "Parenchyma"))

Idents(sobject.sub.fib.v4) = "genotype"
sobject.sub.fib.wt = subset(sobject.sub.fib.v4, idents = "WT")
#3)
save(sobject.sub.fib.wt,file = file.path(dir,datafolder,paste0(project,"_",iterationname,"_WT.RData")))

Idents(sobject.sub.fib.v4) = "timepoint"
sobject.sub.fib.7dpi = subset(sobject.sub.fib.v4, idents = "7dpi")
#4)
save(sobject.sub.fib.7dpi,file = file.path(dir,datafolder,paste0(project,"_",iterationname,"_7dpi.RData")))

rm(sobject.sub.fib.v3, sobject.sub.fib.v4)

Idents(sobject.sub.fib.v2) = "fibtype"
name = paste0(project, iterationname, "_v2_no710_fibtype.eps")
p1 = DimPlot(sobject.sub.fib.v2, label=T)
setEPS()
postscript(file.path(dir,Plotfolder,name))
print(p1)
dev.off()
name = paste0(project, iterationname, "_v2_no710_timepoint.eps")
p2 = DimPlot(sobject.sub.fib.v2, group.by = "timepoint", cols = c("#F8766D", "#DFAF37", "#00BFC4", "#C77CFF"))
setEPS()
postscript(file.path(dir,Plotfolder,name))
print(p2) + NoLegend()
dev.off()

name = paste0(project, iterationname, "_v2_no710_microanatomy.eps")
p3 = DimPlot(sobject.sub.fib.v2, group.by = "microanatomy")
setEPS()
postscript(file.path(dir,Plotfolder,name))
print(p3) + NoLegend()
dev.off()

Idents(sobject.sub.fib.wt.all) = "fibtype"
name = paste0(project, iterationname, "_wt_all_fibtype.eps")
p1 = DimPlot(sobject.sub.fib.wt.all, label=T)
setEPS()
postscript(file.path(dir,Plotfolder,name))
print(p1)
dev.off()
name = paste0(project, iterationname, "_wt_all_timepoint.eps")
p2 = DimPlot(sobject.sub.fib.wt.all, group.by = "timepoint")
setEPS()
postscript(file.path(dir,Plotfolder,name))
print(p2)
dev.off()
```
***~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~***
Redo FindAllMarkers (for all clusters), but exclude cKO (use sobject.sub.fib.wt.all)
NOTE: in later analysis, changed *back* to all genotypes (to reflect genes that determined clusters in the first place) - used code below, but changed to global marker list and from "sobject.sub.fib.wt.all" to "sobject.sub.v2" to "sobject" (still used sobject #12)
***~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~***
First calculate DE genes for every cluster (without meninges)
```{r,results = "hide",eval = FALSE}
Idents(sobject.sub.fib.wt.all) = "fibtype" 
markers_all <- FindAllMarkers(
    object = sobject.sub.fib.wt.all,
    assay = "RNA",  ##Edited to do DEG analysis on RNA counts, not SCT transformed data
    features = rownames(sobject.sub.fib.wt),
    test.use = "MAST",
    only.pos = FALSE, 
    min.pct = 0.05, 
    logfc.threshold = 0.2
)
write.csv(markers_all,file = file.path(dir,"Spreadsheets",paste0(project, "_", iterationname, "_all_markers_wt.csv")))
```

```{r}
column = "fibtype"
#read in a de gene file
markers = read.csv(file.path(dir,"Spreadsheets",paste0(project, "_", iterationname, "_all_markers_wt.csv")),stringsAsFactors = F)

#or non WT
markers = read.csv(file.path(dir,"Spreadsheets",paste0(project, "_", iterationname, "_all_markers.csv")),stringsAsFactors = F)

#Select only the genes that pass thresholds
markers = markers[markers$p_val_adj<pval,]

#pick only positives, or restrict by min/max pct expression using pct1/2
markers = markers[markers$avg_log2FC > lfc,] #make log2FC >0.2 for heatmaps, >0 for volcano plots
markers = markers[markers$pct.1 > minpct & markers$pct.2 < maxpct,]

#If you want, select markers that define a single cluster
if (single){markers <- markers[markers$gene %in% names(table(markers$gene))[table(markers$gene) == 1],] }

table(markers$cluster)

#dotplot for figure
topgenes <- markers %>% group_by(cluster) %>% top_n(3, avg_log2FC) #top 10 genes
#switch cluster name to seurat cluster

# Idents(sobject.sub.fib.wt.all) = "seurat_clusters"
# clusters = levels(sobject.sub.fib.wt.all) 
# Idents(sobject.sub.fib.wt.all) = "fibtype"
# names = levels(sobject.sub.fib.wt.all)
# names(clusters) = names
# topgenes$number = clusters[topgenes$cluster]

#otherwise (if not just WT)
column = "fibtype" #for name consistency
topgenes$number = topgenes$cluster

topgenes = topgenes[order(as.numeric(as.character(topgenes$number))),]

#NOTE BELOW - keeping "wt_only" even if not (to update illustrator)
setEPS()
postscript(file.path(dir,"DotPlots", paste0(project,"_",iterationname, "_wt_only", column,"_",pval,"_ncells",ncells,"dotplot.eps")), width = 9.5, height = 6)
DotPlot(sobject.sub.fib.v2, assay = "RNA", features = unique(c(topgenes$gene)), cols = "RdBu") + theme(axis.text.x = element_text(angle = 90, hjust = 0.95))
dev.off()

setEPS()
postscript(file.path(dir,"DotPlots", paste0(project,"_",iterationname, "_wt_only_WIDE", column,"_",pval,"_ncells",ncells,"dotplot.eps")), width = 12, height = 6)
DotPlot(sobject.sub.fib.v2, assay = "RNA", features = unique(c(topgenes$gene)), cols = "RdBu") + theme(axis.text.x = element_text(angle = 90, hjust = 0.95))
dev.off()
```

Print heatmap to console

```{r}
DoHeatmap(
    object = sobject.sub.fib.wt.all, 
    features = c(topgenes$gene),
    cells = names(newcellnames),
    group.by = column,
    size = 5,
    label = T,
    draw.lines = T
)
```

***~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~***
Compare clusters (heatmap with manual markers)
***~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~***
```{r}
Plotfolder = "Heatmaps"
name = paste0(project, "_", iterationname)

highlight_fib_short = c("Mki67", "Acta2", "Lrrc15", "Fbn2", "Cthrc1", "Fn1", "Postn", "Fn1", "Col1a1", "Col1a2", "Lum", "Col6a1", "Tgfbi", "Dcn")



Idents(sobject.sub.fib.wt) = "fibtype"
fibtypecluster = levels(as.factor(sobject.sub.fib.wt$fibtype))
#Arrange
fibtypecluster_new = c(fibtypecluster[12], fibtypecluster[14], fibtypecluster[6], fibtypecluster[7], fibtypecluster[10], fibtypecluster[2], fibtypecluster[5], fibtypecluster[3])
fibtypecluster_new[1] = "Mural-Pericyte" #fix dash-vs-underscore error
avgexp <- AverageExpression(sobject.sub.fib.wt, return.seurat = TRUE, verbose = TRUE)
levels(avgexp) = fibtypecluster_new

######################################NOTE: in bioRxiv only####################################
#Arrange
#remove pericytes
#fibtypecluster_new = c(fibtypecluster[14], fibtypecluster[6], fibtypecluster[7], fibtypecluster[10], fibtypecluster[2], fibtypecluster[5], fibtypecluster[3]) #remove fibtypecluster[12], pericytes
#noperi = subset(sobject.sub.fib.wt, ident = c("Lama1", "Tmeff2", "Ptgds", "Fn1", "Cd80", "Ghr", "Prolif"))
#avgexp <- AverageExpression(noperi, return.seurat = TRUE, verbose = TRUE)
#levels(avgexp) = fibtypecluster_new
############################################################################

setEPS()
postscript(file.path(dir,Plotfolder, paste0(name, "no_meninges_fib-heatmap_SHORT.eps")), width = 6.5, height = 6)
DoHeatmap(
    object = avgexp, 
    features = highlight_fib_short,
    size = 5,
    label = T,
    raster = F,
    draw.lines = F) + scale_fill_gradientn(colors = rev(RColorBrewer::brewer.pal(n =4, name = "RdBu"))) + theme(text = element_text(size = 17))
dev.off

#------------------------------------------------------------------------------------------#
#unbiased filter for chemokines
#------------------------------------------------------------------------------------------#

######################################NOTE: in bioRxiv only####################################
#chemokines = c("Ccl6", "Cxcl14", "Ccl25", "Cxcl9", "Cxcl10", "Ccl9", "Ccl2", "Ccl7",   "Cxcl2", "Ccl19", "Cxcl1", "Ccl8", "Cxcl12", "Cxcl15", "Cxcl16", "Ccl17", "Ccl28", "Cx3cl1", "Cxcl17", "Cxcl5", "Cxcl13", "Ccl22", "Ccl27a", "Ccl12", "Ccl4") #removed "Ccl11", "Ccl3", "Ccl5", pericyte markers
##########################################################################
chemokines_pericyte = c("Ccl11", "Ccl3", "Ccl5", "Ccl6", "Cxcl14", "Ccl25", "Cxcl9", "Cxcl10", "Ccl9", "Ccl2", "Ccl7",   "Cxcl2", "Ccl17", "Ccl19", "Cxcl1", "Ccl8", "Cxcl12", "Cxcl15", "Cxcl16", "Ccl28", "Cx3cl1", "Cxcl17", "Cxcl5", "Cxcl13", "Ccl22", "Ccl27a", "Ccl12", "Ccl4")

genes = rownames(sobject.sub.fib.wt)
chemokines_all = grep("Ccl|Cxcl|Cx3cl", genes, value = TRUE)
p1 = DotPlot(sobject.sub.fib.wt, chemokines_all, group.by = "fibtype")

chemokines_filtered = as.vector(unique(p1$data[(p1$data$pct.exp > 0.5),3]))
chemokines_threshold = chemokines_pericyte[chemokines_pericyte %in% chemokines_filtered]

#Make heatmaps
setEPS()
postscript(file.path(dir,Plotfolder, paste0(name, "no_meninges_chemokine-heatmap_UNBIASED_FORMATTED.eps")), width = 6.5, height = 6)
DoHeatmap(
    object = avgexp, 
    features = chemokines,
    size = 5,
    label = T,
    raster = F,
    draw.lines = F) + scale_fill_gradientn(colors = rev(RColorBrewer::brewer.pal(n =4, name = "RdBu"))) + theme(text = element_text(size = 17))
dev.off()
```

***GENE SET TESTING**
https://sbc.shef.ac.uk/prostate-bioinformatics/rna-seq-gene-set-testing.nb.html

```{r}
#Note - "old" version performed with meningeal clusters, new version performed without

Plotfolder = "Plots/Fibroblast Subset/GO"

plots = list()
names = list()
i=1

# create list of all genes
all_genes = rownames(sobject.sub.fib.wt)

# create list of DE genes
de_all = read.csv(file.path(dir,"Spreadsheets",paste0(project, "_", iterationname, "_all_markers_no_meninges.csv")),stringsAsFactors = F)
  
group_name = "all"
defile = paste0(project, "_", iterationname, "_cluster-", group_name, "_7-vs-21dpi_markers.csv")
de_7vs21 = read.csv(file.path(dir,"Spreadsheets/Fibroblast Timecourse", defile),stringsAsFactors = F)
de21 = de_7vs21[de_7vs21$avg_log2FC>0,]
de7 = de_7vs21[de_7vs21$avg_log2FC<0,]

defile = paste0(project, "_", iterationname, "_cluster-", group_name, "_wt-vs-col1a2-cKO_markers.csv")
de_wtvscKO = read.csv(file.path(dir,"Spreadsheets/7dpi WT vs cKO", defile),stringsAsFactors = F) #any   spreadsheet with gene symbols or other identifiers
decko = de_wtvscKO[de_wtvscKO$avg_log2FC>0,]
dewt = de_wtvscKO[de_wtvscKO$avg_log2FC<0,]

comparison_delist = list(de7, de21, decko, dewt)

#generate list of fib clusters from de_all (using names)
clusters_named = levels(as.factor(sobject.sub.fib.wt$fibtypecluster))
#sort by ascending value of cluster number
nums_to_sort = c(6, 5, 9, 1, 11, 13, 4, 2)
names(nums_to_sort) = clusters_named
nums_to_sort = sort(nums_to_sort)
# "clusters_named" variable contains cluster number and name; "clusters" variable contains only number
clusters_named = names(nums_to_sort)
clusters = nums_to_sort

#clusters_num contains clusters compared to each other, final "clusters_named" variable will also contain 7 vs. 21dpi and WT vs. cKO (keep track in order to name files differently)
clusters_num = clusters_named
clusters_named = c(clusters_named, "7dpi", "21dpi", "cKO", "WT")

for(cluster in clusters_named){

  #numeric cluster num (no name) for finding in defile
  cluster_find = clusters[which(clusters_named == cluster)]
  
  print(paste0("Iteration: ", cluster))
  #if numeric cluster from cluster comparison, set up
  if(cluster %in% clusters_num){
    name = paste0(project, iterationname, "_cluster_", cluster)
    graphname = paste0("Cluster ", cluster)
    
    de = de_all[which(de_all$cluster == cluster_find),]
  } else{  #if 7vs21dpi or cKOvswt comparisons
    name = paste0(project, iterationname, "_", cluster)
    graphname = cluster
    
    de = comparison_delist[[which(clusters_named == cluster) - 8]]     # get index of delist in "comparison_delist" by subtracting numeric clusters to get remaining 4
  }
  
  sig_de_genes = de[(de$p_val_adj < 0.05),][,1]
  
  #enrichGO
  
  enrich_go <- enrichGO(
    gene= sig_de_genes,
    OrgDb = org.Mm.eg.db,
    keyType = "SYMBOL",
    ont = c("ALL"),
    universe = all_genes,
    qvalueCutoff = 0.05,
    readable=TRUE
  )
  
  empty = FALSE
  num_enriched = length(enrich_go@result$ID)
  if(num_enriched == 0){
    empty = TRUE
    print(paste0(cluster, " is empty"))
    p1 = p2 = p3 = "No terms enriched"
  } else print(paste0(cluster, " enriched: ", num_enriched))
  
  if(empty == FALSE){p1 = dotplot(enrich_go, showCategory = 20, font.size = 8) + ggtitle(graphname) + theme(plot.title = element_text(hjust = 0.5))}

  name1 = paste0(name, "_GO_enrich_dotplot.eps")
  plots[[i]] = p1
  names[[i]] = name1
  i = i+1

  if(empty == FALSE){p2 = barplot(enrich_go, showCategory = 20, font.size = 8) + ggtitle(graphname) + theme(plot.title = element_text(hjust = 0.5))}
  name2 = paste0(name, "_GO_enrich_barplot.eps")
  plots[[i]] = p2
  names[[i]] = name2
  i = i+1
  
  if(empty == FALSE){p3 = dotplot(enrich_go, split="ONTOLOGY", font.size = 8) + facet_grid(ONTOLOGY~., scale="free")  + ggtitle(graphname) + theme(plot.title = element_text(hjust = 0.5))}
  name3 = paste0(name, "_GO_enrich_dotplot_split.eps")
  plots[[i]] = p3
  names[[i]] = name3
  i = i+1
  
  print(paste0("Success: ", cluster, "       i: ", i))
  if(cluster == "Cd80-6"){
    print("Cd80 special graph")
    p4 = dotplot(enrich_go, showCategory = 8, font.size = 12) + ggtitle(graphname) + theme(plot.title = element_text(hjust = 0.5))
    setEPS()
    postscript(file.path(dir,Plotfolder,"Cd80-dotplot-FORMATTED.eps"), height = 5)
    print(p4)
    dev.off()
    setEPS()
    postscript(file.path(dir,Plotfolder,"Cd80-dotplot-LEGEND.eps"))
    print(p4)
    dev.off()
  }
  if(cluster == "Fn1-5"){
    print("Fn1 special graph")
    p4 = dotplot(enrich_go, showCategory = 8, font.size = 12) + ggtitle(graphname) + theme(plot.title = element_text(hjust = 0.5))
    setEPS()
    postscript(file.path(dir,Plotfolder,"Fn1-dotplot-FORMATTED.eps"), height = 5)
    print(p4)
    dev.off()
    setEPS()
    postscript(file.path(dir,Plotfolder,"Fn1-dotplot-LEGEND.eps"))
    print(p4)
    dev.off()
  }
}

for(i in 1:length(plots)){
  plot = plots[[i]]
  name = names[[i]]

  setEPS()
  postscript(file.path(dir,Plotfolder,name))
  print(plot)
  dev.off()
}
```

***~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~***
#Nebulosa plots
```{r}
Plotfolder = "Plots/Fibroblast Subset/Nebulosa"
#use sobject.sub.fib.wt to eliminate osteoblast contribution
pial_genes = c("Lama1", "Col18a1")
arachnoid_genes = c("Ptgds", "Aldh1a2")
dural_genes = c("Alpl", "Foxc2")
gene_lists = list(pial_genes, arachnoid_genes, dural_genes)
names = c("pial_genes", "arachnoid_genes", "dural_genes")
my.pal = palette(brewer.pal(n = 8, name = "YlOrRd"))

for(i in 1:length(gene_lists)){
  genes = gene_lists[[i]]
  list_name = names[i]
  name = paste0(project, "_", iterationname, "_", list_name, "_nebulosa.eps")
  #p1 = plot_density(sobject.sub.fib.wt, genes, joint=T, pal = "magma")
  p1 = Plot_Density_Custom(sobject.sub.fib.wt, genes, joint=T, custom_palette = my.pal) #keep pt size at 1

  setEPS()
  postscript(file.path(dir,Plotfolder,name), width = 14, height = if((length(genes)/3)<=1){3.5}else{7})
  print(p1)
  dev.off()
}
```

```{r}
#parse Ghr fibroblasts

Idents(sobject.sub.fib.wt) = "fibtype"
fibtypecluster = levels(sobject.sub.fib.wt)
fibtypeclusternew = c(fibtypecluster[8], fibtypecluster[4], fibtypecluster[6], fibtypecluster[5], fibtypecluster[1], fibtypecluster[3], fibtypecluster[2], fibtypecluster[7])

sobject.sub.fib.wt@active.ident <- factor(x = sobject.sub.fib.wt@active.ident, levels = fibtypeclusternew)

p1 = VlnPlot(sobject.sub.fib.wt, c("Col1a1", "Cdh18", "Sema3c"), flip=T, stack=T)

Plotfolder = "Plots/figure"
setEPS()
postscript(file.path(dir, Plotfolder, "Ghr_SM.eps"), height = 5)
print(p1)
dev.off()
```

```{r}
# tree using sobject.sub.fib.wt
Idents(sobject.sub.fib.wt) = "fibtype"
sobject.sub.fib.wt= BuildClusterTree(sobject.sub.fib.wt,dims = pcs)
tree = sobject.sub.fib.wt@tools$BuildClusterTree

setEPS()
postscript(file.path(dir,"Trees",paste0(project,"_",iterationname,"_tree_wildtype.eps")))
plot.phylo(tree, use.edge.length = T, direction = "rightwards")
dev.off()
```
