---
title: "Processing"
author: "Nathan Ewing-Crystal"
date: "10/27/2022"
output: html_document
---

#
Seurat information from https://satijalab.org/seurat/articles/spatial_vignette.html
Other code from Leah Dorman

```{r}
#install.packages('Seurat', dependencies=TRUE)
library(Seurat)
#install.packages('ggplot2')
library(ggplot2)
#install.packages('ggpubr')
library(ggpubr)
#install.packages('dplyr')
library(dplyr)
#install.packages('ape')
library(ape)
#install.packages('cowplot')
library(cowplot)
#install.packages('Matrix')
library(Matrix)
#install.packages('EnhancedVolcano')
library(EnhancedVolcano)
#install.packages('knitr')
library(knitr)
#install.packages('readr')
library(readr)
#if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
#BiocManager::install("MAST")
library(MAST)
#install.packages('gplots')
library(gplots)
#BiocManager::install("EnhancedVolcano")
library(EnhancedVolcano)

#BiocManager::install("goseq")
library(goseq)
#BiocManager::install("TxDb.Mmusculus.UCSC.mm10.ensGene")
library(TxDb.Mmusculus.UCSC.mm10.ensGene)

#install.packages("formattable")
library(formattable)

#BiocManager::install("clusterProfiler")
library(clusterProfiler)
#install.packages("ggnewscale")
library(ggnewscale)

#install.packages("remotes")
#remotes::install_github("immunogenomics/presto")
library(presto)
#install.packages("msigdbr")
library(msigdbr)
#BiocManager::install("fgsea")
library(fgsea)
#install.packages("tibble")
library(tibble)
#BiocManager::install("reactome.db")
library(reactome.db)
library(RColorBrewer)
```

*Edit the following code every time* 
and make sure the folders "QC" and "Plots" and "Heatmaps" and "Trees" and "Spreadsheets" and "Data/Seurat" are present in the "dir" folder
```{r}
#Specify your preferred directory for all input + output
# dir="/mnt/DATA/molofsky_lab/newing-crystal/Visium/Seurat/"
dir = "/Users/nathanewing-crystal/Ari Molofsky Lab Dropbox/Ari Molofsky Lab Team Folder/Nathan Ewing-Crystal/RNASeq/Visium/Seurat"
setwd(dir)

#Specify exactly where your seurat files live
datafolder = "Data"

#This name needs to match your project name within the seurat object
project<-"visium" 

#set up folders
QCfolder = "QC"
Plotfolder = "Plots"

#Important genes to determine your cells of interest
igenes = c("Col1a1", "Pdgfra", "Pdgfrb", "Dcn", "Gfap", "Olig2", "Ptprc", "Itgam", "Des")

#make a unique name; maybe this is the celltype you've already subset, or the age you're looking at, etc.
iterationname = "v1"

#Establish cutoffs for heatmaps
pval = 1e-3 #max p-value for significance
lfc = 0.2 #minimum log fold change
minpct = 0.05 #if you want to subset by percent cells in that cluster expressing the gene
maxpct = 1
single = F #should each gene be a marker of a single cluster only
hgenes = 3 #how many genes should be in the heatmap per cluster
ncells = 100 #max # of cells per heatmap column
#column = "sub.cluster" #division you care about

image1 = "A.Rest.2dpi"
image2 = "B.7dpi"
image3 = "C.21dpi"
image4 = "D.21dpi.Deleter" #Note - capture area 4 not analyzed in paper
```
********BELOW CODE PERFORMED ON PRESIDIO SERVER********

#Load individual Visium objects into Seurat
```{r}
filepath <- file.path("/mnt/DATA/molofsky_lab/newing-crystal/Visium/SpaceRanger/NEC_A1_0-2dpi/outs")
vobj_A <- Load10X_Spatial(filepath, filename = "filtered_feature_bc_matrix.h5", assay = "Spatial", slice = "A:Rest/2dpi", filter.matrix = TRUE)

VlnPlot(vobj_A, features = "nCount_Spatial", pt.size = 0.1) + NoLegend()
SpatialFeaturePlot(vobj_A, features = "nCount_Spatial") + theme(legend.position = "right")

vobj_A <- SCTransform(vobj_A, assay = "Spatial", verbose = FALSE)

vobj_A <- RunPCA(vobj_A, assay = "SCT", verbose = FALSE)
vobj_A <- FindNeighbors(vobj_A, reduction = "pca", dims = 1:30)
vobj_A <- FindClusters(vobj_A, verbose = FALSE)
vobj_A <- RunUMAP(vobj_A, reduction = "pca", dims = 1:30)

p1 <- DimPlot(vobj_A, reduction = "umap", label = TRUE)
p2 <- SpatialDimPlot(vobj_A, label = TRUE, label.size = 3)
p1 + p2
```

```{r}
filepath <- file.path("/mnt/DATA/molofsky_lab/newing-crystal/Visium/SpaceRanger/NEC_B1_7dpi/outs")
vobj_B <- Load10X_Spatial(filepath, filename = "filtered_feature_bc_matrix.h5", assay = "Spatial", slice = "B:7dpi", filter.matrix = TRUE)

VlnPlot(vobj_B, features = "nCount_Spatial", pt.size = 0.1) + NoLegend()
SpatialFeaturePlot(vobj_B, features = "nCount_Spatial") + theme(legend.position = "right")

vobj_B <- SCTransform(vobj_B, assay = "Spatial", verbose = FALSE)

vobj_B <- RunPCA(vobj_B, assay = "SCT", verbose = FALSE)
vobj_B <- FindNeighbors(vobj_B, reduction = "pca", dims = 1:30)
vobj_B <- FindClusters(vobj_B, verbose = FALSE)
vobj_B <- RunUMAP(vobj_B, reduction = "pca", dims = 1:30)

p1 <- DimPlot(vobj_B, reduction = "umap", label = TRUE)
p2 <- SpatialDimPlot(vobj_B, label = TRUE, label.size = 3)
p1 + p2
```

```{r}
filepath <- file.path("/mnt/DATA/molofsky_lab/newing-crystal/Visium/SpaceRanger/NEC_C1_21dpi/outs")
vobj_C <- Load10X_Spatial(filepath, filename = "filtered_feature_bc_matrix.h5", assay = "Spatial", slice = "C:21dpi", filter.matrix = TRUE)

VlnPlot(vobj_C, features = "nCount_Spatial", pt.size = 0.1) + NoLegend()
SpatialFeaturePlot(vobj_C, features = "nCount_Spatial") + theme(legend.position = "right")

##some spots have 0 UMIs
# AACGTCATCCGGCTTG-1 AGCGACTTTGAAGACA-1 CTTTATCCGACGCATG-1 TGGATGGCATCTTGGA-1 
#                 70                339               1082               1982 

#filter

vobj_C_subset <- subset(vobj_C, cells = which(vobj_C$nCount_Spatial != 0))
vobj_C_subset <- SCTransform(vobj_C_subset, assay = "Spatial", verbose = FALSE)

vobj_C_subset <- RunPCA(vobj_C_subset, assay = "SCT", verbose = FALSE)
vobj_C_subset <- FindNeighbors(vobj_C_subset, reduction = "pca", dims = 1:30)
vobj_C_subset <- FindClusters(vobj_C_subset, verbose = FALSE)
vobj_C_subset <- RunUMAP(vobj_C_subset, reduction = "pca", dims = 1:30)

p1 <- DimPlot(vobj_C_subset, reduction = "umap", label = TRUE)
p2 <- SpatialDimPlot(vobj_C_subset, label = TRUE, label.size = 3)
p1 + p2
```

#Note - capture area 4 not analyzed in paper
```{r}
filepath <- file.path("/mnt/DATA/molofsky_lab/newing-crystal/Visium/SpaceRanger/NEC_D1_21dpi_deleter/outs")
vobj_D <- Load10X_Spatial(filepath, filename = "filtered_feature_bc_matrix.h5", assay = "Spatial", slice = "D:21dpi Deleter", filter.matrix = TRUE)

VlnPlot(vobj_D, features = "nCount_Spatial", pt.size = 0.1) + NoLegend()
SpatialFeaturePlot(vobj_D, features = "nCount_Spatial") + theme(legend.position = "right")

vobj_D <- SCTransform(vobj_D, assay = "Spatial", verbose = FALSE)

vobj_D <- RunPCA(vobj_D, assay = "SCT", verbose = FALSE)
vobj_D <- FindNeighbors(vobj_D, reduction = "pca", dims = 1:30)
vobj_D <- FindClusters(vobj_D, verbose = FALSE)
vobj_D <- RunUMAP(vobj_D, reduction = "pca", dims = 1:30)

p1 <- DimPlot(vobj_D, reduction = "umap", label = TRUE)
p2 <- SpatialDimPlot(vobj_D, label = TRUE, label.size = 3)
p1 + p2
```
#
#Merge
```{r}
vobj_A$orig.ident <- "Rest, 2dpi"
vobj_B$orig.ident <- "7dpi"
vobj_C_subset$orig.ident <- "21dpi"
vobj_D$orig.ident <- "21dpi Deleter"

vobj.merge_ab <- merge(vobj_A, vobj_B)
vobj.merge_abc <- merge(vobj.merge_ab, vobj_C_subset)
vobj.merge <- merge(vobj.merge_abc, vobj_D)

VlnPlot(vobj.merge, features = "nCount_Spatial", pt.size = 0.1) + NoLegend()
SpatialFeaturePlot(vobj.merge, features = "nCount_Spatial") + theme(legend.position = "right")

DefaultAssay(vobj.merge) <- "SCT"
VariableFeatures(vobj.merge) <- c(VariableFeatures(vobj_A), VariableFeatures(vobj_B), VariableFeatures(vobj_C_subset), VariableFeatures(vobj_D))
vobj.merge <- RunPCA(vobj.merge, verbose = FALSE)
vobj.merge <- FindNeighbors(vobj.merge, dims = 1:30)
vobj.merge <- FindClusters(vobj.merge, verbose = FALSE)
vobj.merge <- RunUMAP(vobj.merge, dims = 1:30)
```

```{r}
DimPlot(vobj.merge, reduction = "umap", group.by = c("ident", "orig.ident"))
```






Function to print multiple graphs: 
```{r}
PrintSeuratGraph = function(namecard = "a",seurat_object = vobj.merge,graphtype = "feature",feature = NULL,group = NULL,split=NULL,cellnames=NULL, label=FALSE, images=NULL){
  if (!is.null(cellnames)){
    Idents(seurat_object) = cellnames[1]
    cells = colnames(seurat_object)[Idents(seurat_object) %in% cellnames[2:length(cellnames)]]} 
  else {cells = cellnames}
  if (graphtype == "feature"){
    graph = FeaturePlot(seurat_object,features = feature,split.by = split, cells = cells,cols = c("lightyellow","darkred"))
  }
  if (graphtype == "violin"){
    graph = VlnPlot(seurat_object,features = feature, pt.size = 0.1, idents = cellnames[2:length(cellnames)],group.by = group, split.by = split)
  }
  if (graphtype == "dim"){
    graph = DimPlot(seurat_object,cells = cells, group.by = group, split.by = split, label=label)
    
  }
  if (graphtype == "spatialfeature"){
    graph = SpatialFeaturePlot(seurat_object,features = feature,images=images)
  }
   if (graphtype == "spatialdim"){
    graph = SpatialDimPlot(seurat_object,cells = cells, group.by = group, label=label)
  }
  name = paste0(feature,"_",graphtype,namecard,".eps")
  graph
  setEPS()
  postscript(file.path(dir,Plotfolder,name))
  print(graph)
  dev.off()
}
```

Block to print multiple graphs: 
```{r}
name = paste0(project,iterationname)
genes = igenes
genes = genes[genes %in% rownames(GetAssayData(vobj.merge,slot = "data"))]

for(feature in genes){
  PrintSeuratGraph(namecard = name,graphtype = "feature",feature = feature)
  PrintSeuratGraph(namecard = name,graphtype = "spatialfeature",feature = feature)
}

#dim plots for clustering
PrintSeuratGraph(namecard = name,graphtype = "dim", label=TRUE)
PrintSeuratGraph(namecard = paste0(name, "_grouped") ,graphtype = "dim", group = "orig.ident")
PrintSeuratGraph(namecard = name,graphtype = "spatialdim", legend=FALSE)

```

```{r}
#Find all markers

markers_all = FindAllMarkers(
   object = vobj.merge,
   assay = "Spatial",
   features = rownames(vobj.merge),
   test.use = "MAST", 
   only.pos = FALSE, 
   min.pct = 0.05, 
   logfc.threshold = 0.2)

write.csv(markers_all, paste0(dir, "Tables/markers_all.csv"))
```

```{r}
save(vobj.merge,file = file.path(dir, datafolder,paste0(project,"_vobj.merge_",iterationname,".RData")))

#CHANGE DATE AS NECESSARY
save.image(file = file.path(dir, datafolder, "workspace.20221028.Rdata"))
```

********BELOW CODE PERFORMED locally********

```{r}
filename = "/Users/nathanewing-crystal/Dropbox (Ari Molofsky Lab)/Ari Molofsky Lab Team Folder/Nathan Ewing-Crystal/RNASeq/Visium/Seurat/Data/visium_vobj.merge_v1.RData"
load(file.path(filename))
```

```{r}
p1 <- SpatialDimPlot(vobj.merge.sub10, images = image1, pt.size.factor = 0.5) + theme(legend.position = "none")
p2 <- SpatialDimPlot(vobj.merge.sub10, images = image2, pt.size.factor = 0.5) + theme(legend.position = "none")
p3 <- SpatialDimPlot(vobj.merge.sub10, images = image3, pt.size.factor = 0.5) + theme(legend.position = "none")
p4 <- SpatialDimPlot(vobj.merge.sub10, images = image1) + theme(legend.position = "none")
p5 <- SpatialDimPlot(vobj.merge.sub10, images = image2) + theme(legend.position = "none")
p6 <- SpatialDimPlot(vobj.merge.sub10, images = image3) + theme(legend.position = "none")
Plotfolder = "Plots"
setEPS()
postscript(file.path(dir, Plotfolder, "small_dots_spatial.eps"))
p1+p2+p3
dev.off()
setEPS()
postscript(file.path(dir, Plotfolder, "big_dots_spatial.eps"))
p4+p5+p6
dev.off()
```

#Parse fibroblast expressing clusters (differences over time) - replicate cloupe analysis
```{r}
Plotfolder = "Plots/fibroblast_parsing"

col_clusters = c(3, 8, 10, 13, 17, 18)
for(cluster in col_clusters){
    name = paste0("cluster_", cluster, ".eps")

    p6 <- DimPlot(vobj.merge, cells.highlight=CellsByIdentities(object = vobj.merge,idents = cluster)) + theme(legend.position = "top")

    p7 <- DimPlot(subset(vobj.merge, subset = orig.ident %in% c("Rest, 2dpi", "7dpi", "21dpi")), cells.highlight=CellsByIdentities(object = vobj.merge,idents = cluster), split.by = "orig.ident") + theme(legend.position = "none")
  #change order of factor levels --> plot from 2dpi to 21dpi
  p7$data$orig.ident <- factor(x = p7$data$orig.ident, levels = c("Rest, 2dpi", "7dpi", "21dpi"))

  p8 <- SpatialDimPlot(vobj.merge, cells.highlight=CellsByIdentities(object = vobj.merge,idents = cluster), images=c(image1, image2, image3)) + theme(legend.position = "none")
  
  setEPS()
  postscript(file.path(dir,Plotfolder,name))
  print(
    ggarrange(ggarrange(p5, p6, ncol=2, nrow=1),
              p7, p8, ncol=1, nrow=3))
  dev.off()
}

#parse by timepoint
Idents(vobj.merge) = "seurat_clusters"
for(cluster in col_clusters){
  name = paste0("cluster_", cluster, "_timepoint.eps")
  p12 <- DimPlot(subset(vobj.merge, subset = orig.ident %in% c("Rest, 2dpi", "7dpi", "21dpi")), cells.highlight=CellsByIdentities(object = vobj.merge,idents = cluster), split.by = "timepoint") + theme(legend.position = "none")
  #change order of factor levels --> plot from 2dpi to 21dpi
  p12$data$timepoint <- factor(x = p12$data$timepoint, levels = c("Rest", "2dpi", "7dpi", "21dpi"))
  setEPS()
  postscript(file.path(dir,Plotfolder,name), height = 2.5)
  print(p12)
  dev.off()
}
```

#Subcluster cluster 10
```{r}
vobj.merge.sub10 <- FindSubCluster(
  vobj.merge,
  10,
  graph.name = "SCT_snn",
  subcluster.name = "sub.cluster",
  resolution = 0.5,
  algorithm = 1
)

vobj.merge.sub10 <- SetIdent(vobj.merge.sub10, value = vobj.merge.sub10@meta.data$sub.cluster)
DimPlot(vobj.merge.sub10, label = TRUE)

#Subset to new seurat object with only cluster 10
vobj.merge.sub10 <- SetIdent(vobj.merge.sub10, value = vobj.merge.sub10@meta.data$seurat_clusters)
sub10 <- subset(vobj.merge.sub10, idents = 10)

vobj.merge.sub10 <- SetIdent(vobj.merge.sub10, value = vobj.merge.sub10@meta.data$sub.cluster)
subclusters <- c("10_0", "10_1", "10_2", "10_3", "10_4")

sub10 <- SetIdent(sub10, value = sub10@meta.data$sub.cluster)

for(subcluster in subclusters){
    name = paste0("cluster_", subcluster, ".eps")

    p6 <- DimPlot(vobj.merge.sub10, cells.highlight=CellsByIdentities(object = vobj.merge.sub10,idents = subcluster)) + theme(legend.position = "top")

    p7 <- DimPlot(subset(vobj.merge.sub10, subset = orig.ident %in% c("Rest, 2dpi", "7dpi", "21dpi")), cells.highlight=CellsByIdentities(object = vobj.merge.sub10,idents = subcluster), split.by = "orig.ident") + theme(legend.position = "none")
  #change order of factor levels --> plot from 2dpi to 21dpi
  p7$data$orig.ident <- factor(x = p7$data$orig.ident, levels = c("Rest, 2dpi", "7dpi", "21dpi"))

  p8 <- SpatialDimPlot(vobj.merge.sub10, cells.highlight=CellsByIdentities(object = vobj.merge.sub10,idents = subcluster), images=c(image1, image2, image3)) + theme(legend.position = "none")
  
  setEPS()
  postscript(file.path(dir,Plotfolder,name))
  print(
    ggarrange(ggarrange(p5, p6, ncol=2, nrow=1),
              p7, p8, ncol=1, nrow=3))
  dev.off()
}
```

***COMPARE FIBROBLASTS 7 vs. 21dpi***

#Get subset markers for 21dpi vs. 7dpi fibroblasts
```{r}
#scale RNA assay data

DefaultAssay(vobj.merge.sub10) <- "Spatial"
vobj.merge.sub10 <- NormalizeData(vobj.merge.sub10)
vobj.merge.sub10 <- ScaleData(vobj.merge.sub10)

vobj.merge.sub10 <- SetIdent(vobj.merge.sub10, value = vobj.merge.sub10@meta.data$sub.cluster)
cluster1 = "8"
cluster2 = "10_0"

#Find markers based on scaled Spatial assay
markers_7vs21_fib = FindMarkers(
  object = vobj.merge.sub10,
  assay = "Spatial",
  ident.1 = cluster1,
  ident.2 = cluster2,
  test.use = "MAST",
  only.pos = FALSE,
  min.pct = 0.05,
  logfc.threshold = 0.2)

write.csv(markers_7vs21_fib, paste0(dir, "/Tables/markers_7-vs-21-fib.csv"))
```
*Volcano Plot*
7 vs. 21dpi

Set your parameters
```{r}
#Minimum fold change (i.e. 1.15 = 15% increase)
minfc = 1.15
#Max adj. p value
alpha = 0.05
#Genes to highlight
ngenes = 50

defile = "markers_7-vs-21-fib.csv"

#Comparison being made (must match file names)
comparison = "markers_7-vs-21-fib"

newlist=list(de)
de = read.csv(file.path(dir,"Tables", defile),stringsAsFactors = F) #any   spreadsheet with gene symbols or other identifiers
colnames(de)[1] = "Gene"

#NOTE: generated from "DEG_analysis" script
fc = de
fc = fc[!is.na(fc$avg_log2FC),]
colorkeysdown = fc$Gene[fc$avg_log2FC < -log2(minfc) & fc$p_val_adj < alpha]
colorkeysup = fc$Gene[fc$avg_log2FC > log2(minfc) & fc$p_val_adj < alpha]

#Either highlight specific genes or pick the top genes in colorkeysup/down
top = fc[fc$p_val_adj<alpha,]
top = top[order(top$avg_log2FC),"Gene"]
highlight_top = c(head(top,ngenes),tail(top,ngenes))
highlight_immune = c("Cxcl12",  "Ccl3", "Ccl2", "Ccl4", "Ifit3", "Itgam", "C3", "Ptgds", "Il1rn", "Irf7", "Ifitm2", "Ifit3b", "Ifit1", "Ifitm3", "Ifit2","Ccl12", "Ccl9", "Ccl7", "Csf1", "H2-Aa", "Spp1")
highlight_fib = c("Mki67", "Cthrc1", "Lrrc15", "Acta2", "Dcn",  "Col1a2", "Lum", "Itgav", "Col6a1", "Col1a1", "Col8a1")
genelists = list(highlight_immune, highlight_fib, highlight_top)
names_genelists = c("immune", "fibrosis", "top")
size_labels = c(4, 4, 2)

allcolors = rep("darkgrey",length(fc$Gene))
names(allcolors) = fc$Gene
allcolors[names(allcolors) %in% colorkeysdown] = "red"
allcolors[names(allcolors) %in% colorkeysup]= "blue"
names(allcolors)[allcolors == "yellow"] = "labelled"
names(allcolors)[allcolors == "blue"] = "Enriched  21dpi"
names(allcolors)[allcolors == "darkgrey"] = "-"
names(allcolors)[allcolors == "red"] = "Enriched 7dpi"
boxed_labels = c(T, T, F)
  
#manual label
name = "Fibroblasts 7 vs. 21dpi"
range = max(max(fc$avg_log2FC), min(fc$avg_log2FC))
if(is.infinite(range)){range = 150}

for(i in 1:length(names_genelists)){
  setEPS()
  postscript(file.path(dir,"Plots/fibroblast_parsing",paste0("Volcano_",comparison,"_", names_genelists[i], ".eps")))
  print(EnhancedVolcano(fc,
              lab = fc$Gene,
              x = 'avg_log2FC',
              y = 'p_val_adj',
              xlim = c(-range, range),
              title = name,
              drawConnectors = T,
              arrowheads = F,
              legendPosition = 'bottom',
              pCutoff = alpha,
              FCcutoff = log2(minfc),
              selectLab = genelists[[i]],
              labSize = size_labels[i],
              pointSize = 0.5,
              col=c('black', 'black', 'black', 'red3'),
              colCustom = allcolors,
              gridlines.major = F,
              gridlines.minor = F,
              colAlpha = 1,
              max.overlaps = 50))
  dev.off()
  
  setEPS()
  postscript(file.path(dir,"Plots/fibroblast_parsing",paste0("Volcano_",comparison,"_", names_genelists[i], "_large-text.eps")))
  print(EnhancedVolcano(fc,
              lab = fc$Gene,
              x = 'avg_log2FC',
              y = 'p_val_adj',
              xlim = c(-range, range),
              title = name,
              drawConnectors = T,
              arrowheads = F,
              legendPosition = 'bottom',
              pCutoff = alpha,
              FCcutoff = log2(minfc),
              selectLab = genelists[[i]],
              labSize = 6,
              pointSize = 0.5,
              col=c('black', 'black', 'black', 'red3'),
              colCustom = allcolors,
              gridlines.major = F,
              gridlines.minor = F,
              colAlpha = 1,
              max.overlaps = 50,
              boxedLabels = boxed_labels[i]))
  dev.off()
}

```

```{r}
#fibrosis 7 vs 21dpi final figure
range_min = min(fc$avg_log2FC)
range_max = max(fc$avg_log2FC)

setEPS()
postscript(file.path(dir,"Plots/fibroblast_parsing",paste0("Volcano_",comparison,"_", names_genelists[i], "_large-text_FINAL.eps")))
print(EnhancedVolcano(fc,
              lab = NA,
              x = 'avg_log2FC',
              y = 'p_val_adj',
              xlim = c(range_min, range_max),
              title = name,
              drawConnectors = T,
              arrowheads = F,
              legendPosition = 'bottom',
              pCutoff = alpha,
              FCcutoff = log2(minfc),
              selectLab = genelists[[i]],
              labSize = 6,
              pointSize = 0.5,
              col=c('black', 'black', 'black', 'red3'),
              colCustom = allcolors,
              gridlines.major = F,
              gridlines.minor = F,
              colAlpha = 1,
              max.overlaps = 50, min.segment.length = 0.05))
dev.off()
```

***Subset anatomical region - split rest and 2dpi***
```{r}
# Use SpatialDimPlots to visualize what to remove
coords_rest = which(vobj.merge.sub10@images$A.Rest.2dpi@coordinates$imagerow < 7000)
coords_2dpi = which(vobj.merge.sub10@images$A.Rest.2dpi@coordinates$imagerow > 7000)
rest_2dpi_spot_names = rownames(vobj.merge.sub10@meta.data[which(vobj.merge.sub10$orig.ident == "Rest, 2dpi"),])
rest_spot_names = rest_2dpi_spot_names[coords_rest]
x2dpi_spot_names = rest_2dpi_spot_names[coords_2dpi]

SpatialDimPlot(vobj.merge.sub10, images = "A.Rest.2dpi", cells.highlight = rest_spot_names)
SpatialDimPlot(vobj.merge.sub10, images = "A.Rest.2dpi", cells.highlight = x2dpi_spot_names)

#new timepoint variable, named by cell
timepoint = vobj.merge.sub10@meta.data$orig.ident
names(timepoint) = rownames(vobj.merge.sub10@meta.data)
#21dpi deleter --> 21dpi
timepoint[which(timepoint == "21dpi Deleter")] = "21dpi"
#rest and 2dpi subset
timepoint[rest_spot_names] = "Rest"
timepoint[x2dpi_spot_names] = "2dpi"


vobj.merge.sub10 <- AddMetaData(vobj.merge.sub10, timepoint, col.name = "timepoint")
Idents(vobj.merge.sub10) <- "timepoint"    

SpatialDimPlot(vobj.merge.sub10)

#PLOT NEW GRAPHS DIVIDED BY TIMEPOINT
Plotfolder = "Plots/timepoint_parsed"
name = paste0(project,iterationname, "_timepoint_parsed_")

#dim plots for clustering
PrintSeuratGraph(seurat_object = vobj.merge.sub10, namecard = paste0(name, "_grouped") ,graphtype = "dim", group = "timepoint")
PrintSeuratGraph(seurat_object = vobj.merge.sub10, namecard = name,graphtype = "spatialdim")

Idents(vobj.merge.sub10) = "seurat_clusters"
p1 = DimPlot(vobj.merge.sub10, split.by = "timepoint")
p1$data$timepoint <- factor(x = p1$data$timepoint, levels = c("Rest", "2dpi", "7dpi", "21dpi"))
name = "_dimvisiumv1_split_wide.eps"

setEPS()
postscript(file.path(dir,Plotfolder,name), width = 10, height = 5)
print(p1)
dev.off()

name = "_dimvisiumv1_split_combined.eps"

p1 <- p1 + theme(legend.position = "none")
p2 <- DimPlot(vobj.merge.sub10)
p3 <- SpatialDimPlot(vobj.merge.sub10, images = image1) + theme(legend.position = "none")
p4 <- SpatialDimPlot(vobj.merge.sub10, images = image2) + theme(legend.position = "none")
p5 <- SpatialDimPlot(vobj.merge.sub10, images = image3) + theme(legend.position = "none")
p6 = p3 + p4 + p5

setEPS()
postscript(file.path(dir,Plotfolder,name), height = 10)
ggarrange(ggarrange(
 p2, p1, nrow = 2), p6, nrow = 2)
dev.off()

```

***ADD MODULE SCORE***
```{r}

Plotfolder = "Plots/module_score"
# proliferation signature from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2819333/#!po=73.5294
#additions from https://www.nature.com/articles/nrc1802#:~:text=The%20proliferation%20signature%20is%20now,and%20cyclin%20B1%2C%20respectively).
features_prolif = list(c("Ccnb1", "Ccne1", "Ccnd1", "E2f1", "Tfdp1", "Cdkn2b", "Cdkn1a", "Plk4", "Wee1", "Aurkb", "Bub1", "Chek1", "Prim1", "Top2a", "Cks1", "Rad51l1", "Shc", "Racgap1", "Cbx1", "Mki67", "Mybl2", "Bub1", "Plk1"))

#below are from MD stromal in vitro experiments - took all genes where log2FC > 0, q < 0.05
features_tgfb = read.csv(file.path(dir,"Data/Other/tgfb-vs-pbs.csv"),stringsAsFactors = F, header=F) 

#ADD MODULE SCORES

vobj.merge.sub10 <- AddModuleScore(
  vobj.merge.sub10,
  features_prolif,
  pool = NULL,
  nbin = 24,
  ctrl = 100,
  name = "proliferation_score",
  seed = 1,
  search = TRUE,
)
#note - added as "proliferation_score1"

vobj.merge.sub10 <- AddModuleScore(
  vobj.merge.sub10,
  features_tgfb,
  pool = NULL,
  nbin = 24,
  ctrl = 100,
  name = "tgfb_score",
  seed = 1,
  search = TRUE,
)
#note - added as "tgfb_score1"

#PREPARE TO GRAPH
Idents(vobj.merge.sub10) = "fib.timepoint"
my_levels <- c("7dpi", "21dpi")
my_levels = c("3", "6", "12", "21", "11", "20", "0", "17", "16", "9", "19", "5", "2", "4", "10_1", "22", "10_4", "1", "18", "7dpi", "21dpi", "14", "13", "23", "15", "10_2", "10_3", "7")
levels(vobj.merge.sub10) = my_levels

#------------PROLIF----------

setEPS()
postscript(file.path(dir, Plotfolder, "prolif.eps"), width = 3.5, height = 4)
VlnPlot(vobj.merge.sub10, idents = c("7dpi", "21dpi"), "proliferation_score1")
dev.off()

setEPS()
postscript(file.path(dir, Plotfolder, "spatial_prolif.eps"))
SpatialFeaturePlot(vobj.merge.sub10, "proliferation_score1", images = c(image1, image2, image3))
dev.off()

#EQUALIZE SCALE BETWEEN 3 SECTIONS
rng = range(vobj.merge.sub10$proliferation_score1) #a range to have the same min and max for both plots

myPalette <- colorRampPalette(rev(brewer.pal(11, "Spectral")))
sc <- scale_fill_gradientn(colours = myPalette(100), limits = c(rng[1], rng[2]), breaks = c(-0.1, 0.0, 0.1, 0.2, 0.3, 0.4, 0.5))

p1 = SpatialFeaturePlot(vobj.merge.sub10, "proliferation_score1", images = c(image1, image2, image3))

setEPS()
postscript(file.path(dir, Plotfolder, "spatial_prolif_normalized_scale.eps"))
p1[[1]] + sc + p1[[2]] + sc + p1[[3]] + sc
dev.off()

#------------TGFb----------

setEPS()
postscript(file.path(dir, Plotfolder, "tgfb.eps"), width = 3.5, height = 4)
VlnPlot(vobj.merge.sub10, idents = c("7dpi", "21dpi"), "tgfb_score1")
dev.off()


setEPS()
postscript(file.path(dir, Plotfolder, "spatial_tgfb.eps"))
SpatialFeaturePlot(vobj.merge.sub10, "tgfb_score1", images = c(image1, image2, image3))
dev.off()

#EQUALIZE SCALE BETWEEN 3 SECTIONS
rng = range(vobj.merge.sub10$tgfb_score1) #a range to have the same min and max for both plots

myPalette <- colorRampPalette(rev(brewer.pal(11, "Spectral")))
sc <- scale_fill_gradientn(colours = myPalette(100), limits = c(rng[1], rng[2]), breaks = c(0.0, 0.1, 0.2, 0.3, 0.4, 0.5))

p1 = SpatialFeaturePlot(vobj.merge.sub10, "tgfb_score1", images = c(image1, image2, image3))

setEPS()
postscript(file.path(dir, Plotfolder, "spatial_tgfb_normalized_scale.eps"))
p1[[1]] + sc + p1[[2]] + sc + p1[[3]] + sc
dev.off()

```

#GRAPHS FOR FIGURE
```{r}
Plotfolder = "Plots/figure"
name = "Col1a1_feature.eps"

#EQUALIZE SCALE BETWEEN 3 SECTIONS
rng = range(vobj.merge.sub10@assays$Spatial["Col1a1",]) #a range to have the same min and max for both plots

myPalette <- colorRampPalette(rev(brewer.pal(11, "Spectral")))
sc <- scale_fill_gradientn(colours = myPalette(100), limits = c(rng[1], rng[2]), breaks = c(0, 2, 4))

p1 = SpatialFeaturePlot(vobj.merge.sub10, "Col1a1", images = c(image1, image2, image3))

setEPS()
postscript(file.path(dir, Plotfolder, name))
p1[[1]] + sc + p1[[2]] + sc + p1[[3]] + sc
dev.off()

options(ggrepel.max.overlaps = Inf)
Idents(vobj.merge.sub10) = "sub.cluster"

p1 <- SpatialDimPlot(vobj.merge.sub10, images = image1, label=T, repel = T, label.size = 2) + theme(legend.position = "none")
p2 <- SpatialDimPlot(vobj.merge.sub10, images = image2, label=T, repel = T, label.size = 2) + theme(legend.position = "none")
p3 <- SpatialDimPlot(vobj.merge.sub10, images = image3, label=T, repel = T, label.size = 2) + theme(legend.position = "none")

setEPS()
postscript(file.path(dir, Plotfolder, "labeled_umap.eps"))
p1 + p2 + p3
dev.off()

Idents(vobj.merge.sub10) = "sub.cluster"
markers = FindAllMarkers(
  object = vobj.merge.sub10,
  assay = "Spatial",
  test.use = "MAST",
  only.pos = FALSE,
  min.pct = 0.05,
  logfc.threshold = 0.2)

write.csv(markers, paste0(dir, "/Tables/markers_all_subcluster.csv"))

defile = "markers_all_subcluster.csv"
markers = read.csv(file.path(dir,"Tables", defile),stringsAsFactors = F) #any   spreadsheet with gene symbols or other identifiers

#Select only the genes that pass thresholds
markers = markers[markers$p_val_adj<pval,]

#pick only positives, or restrict by min/max pct expression using pct1/2
markers = markers[markers$avg_log2FC > lfc,] #make log2FC >0.2 for heatmaps, >0 for volcano plots
markers = markers[markers$pct.1 > minpct & markers$pct.2 < maxpct,]

table(markers$cluster)

#dotplot for figure
topgenes <- markers %>% group_by(cluster) %>% top_n(3, avg_log2FC) #top 10 genes
#switch cluster name to seurat cluster

#ORDER
topgenes$cluster2 = gsub("_", ".", topgenes$cluster)

topgenes = topgenes[order(as.numeric(as.character(topgenes$cluster2))),]
levels(vobj.merge.sub10) = unique(topgenes$cluster)

setEPS()
postscript(file.path(dir,Plotfolder, paste0(project,"_",iterationname, "dotplot.eps")), width = 12.5)
DotPlot(vobj.merge.sub10, assay = "Spatial", features = unique(c(topgenes$gene)), cols = "RdBu") + theme(axis.text.x = element_text(angle = 90, hjust = 0.95))
dev.off()

conserved_fib_genes = c("Pdgfra", "Col1a1", "Col5a1", "Col6a1", "Col7a1", "Dcn")

setEPS()
postscript(file.path(dir,Plotfolder, paste0(project,"_",iterationname, "fibroblast_genes_dotplot.eps")), height = 6, width = 6)
DotPlot(vobj.merge.sub10, assay = "Spatial", features = conserved_fib_genes, cols = "RdBu") + theme(axis.text.x = element_text(angle = 90, hjust = 0.95))
dev.off()

setEPS()
postscript(file.path(dir,Plotfolder, paste0(project,"_",iterationname, "Col1a1_violin.eps")), width = 9, height = 7)
VlnPlot(vobj.merge.sub10, "Col1a1")
dev.off()

```
***MAP snRNASeq clusters via module score***
```{r}
Plotfolder = "Plots/module_score/snRNASeq Mapping"

Idents(vobj.merge.sub10) = "sub.cluster"
name = paste0(project, iterationname)
  
#Map sn clusters (NEC127) onto Visium
old_markers_path = "/Users/nathanewing-crystal/Ari Molofsky Lab Dropbox/Ari Molofsky Lab Team Folder/Nathan Ewing-Crystal/RNASeq/Lesion snRNASeq/seurat/Spreadsheets/lesion_snRNASeq_fib_reclustered_all_markers.csv"
old_markers = read.csv(old_markers_path,stringsAsFactors = F)

old_clusters = levels(factor(old_markers$cluster))
old_cluster_names = c("Meninges_1", "Lama1", "Tmeff2", "Meninges_2", "Ptgds", "Fn1", "Cd80", "Myeloid", "Meninges_3", "Ghr", "Nrxn1-3", "Mural_Pericyte", "Meninges_4", "Prolif", "Meninges_5")
names(old_clusters) = old_cluster_names
old_clusters = old_clusters[-c(8,11)]
old_cluster_names = old_cluster_names[-c(8,11)]

plots = list()

for(i in 1:length(old_clusters)){
  old_cluster = old_clusters[i]
  cluster_name = old_cluster_names[i]
  old_genes = old_markers[old_markers$cluster == old_cluster & old_markers$avg_log2FC > 0,]$gene
  #convert  into a list with length=1
  old_genes_list = list()
  old_genes_list[[1]] = old_genes
  vobj.merge.sub10 <- AddModuleScore(
    vobj.merge.sub10,
    old_genes_list,
    pool = NULL,
    nbin = 24,
    ctrl = 100,
    name = paste0(cluster_name, "_score"),
    seed = 1,
    search = TRUE)
    
    feature_name = paste0(cluster_name, "_score1")
    print(VlnPlot(vobj.merge.sub10, feature_name))
    #PrintSeuratGraph(seurat_object = vobj.merge.sub10, namecard = name,graphtype = "violin",feature = feature_name,group = "sub.cluster")
}
```
```{r}
#for figure

p1 = VlnPlot(vobj.merge.sub10, c("Lama1_score1", "Ptgds_score1", "Tmeff2_score1"), stack=T, same.y.lims=T)
p2 = VlnPlot(vobj.merge.sub10, c("Prolif_score1", "Fn1_score1", "Ghr_score1", "Cd80_score1", "Lama1_score1", "Ptgds_score1", "Tmeff2_score1"), stack=T, same.y.lims=T)
p3 = VlnPlot(vobj.merge.sub10, c("Prolif_score1", "Fn1_score1", "Ghr_score1", "Cd80_score1", "Lama1_score1", "Ptgds_score1", "Tmeff2_score1"), stack=T, same.y.lims=T, flip = T)
p4 = DotPlot(vobj.merge.sub10, assay = "Spatial", features = c("Prolif_score1", "Fn1_score1", "Ghr_score1", "Cd80_score1", "Lama1_score1", "Ptgds_score1", "Tmeff2_score1"), cols = "RdBu") + theme(axis.text.x = element_text(angle = 90, hjust = 0.95))  + coord_flip()
  #not using


ordered = c("0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10_0", "10_1", "10_2", "10_3", "10_4", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23")
p1$data$ident <- factor(x = p1$data$ident, levels = ordered)
p2$data$ident <- factor(x = p2$data$ident, levels = ordered)
p3$data$ident <- factor(x = p3$data$ident, levels = ordered)
p4$data$id <- factor(x = p4$data$id, levels = ordered)

setEPS()
postscript(file.path(dir,Plotfolder, paste0(project,"_",iterationname, "snseq_mapping_meningeal")), height = 6, width = 6)
p1
dev.off()

setEPS()
postscript(file.path(dir,Plotfolder, paste0(project,"_",iterationname, "snseq_mapping_all")), height = 5.25, width = 6)
p2
dev.off()

setEPS()
postscript(file.path(dir,Plotfolder, paste0(project,"_",iterationname, "snseq_mapping_all_FLIP")), height = 4.75, width = 13)
p3 + theme(axis.text.x = element_text(angle = 90))
dev.off()
```

