---
atitle: "10_Immune_subset"
output: html_document
date: "2023-05-30"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(Seurat)
library(ggplot2)
library(dplyr)
library(ape)
library(cowplot)
library(Matrix)
library(MAST)
#install.packages('EnhancedVolcano')
library(EnhancedVolcano)

#BiocManager::install("Nebulosa")
library("Nebulosa")

library(SeuratWrappers)
library(presto)
```

#Load T seurat object
```{r}
load(file.choose())
```

```{r}
#Specify your preferred directory for all input + output
dir = "/Users/nathanewing-crystal/Ari Molofsky Lab Dropbox/Ari Molofsky Lab Team Folder/Nathan Ewing-Crystal/RNASeq/Lesion snRNASeq 2 Tgfb/seurat"
#Specify exactly where your seurat files live
datafolder = "Data/Seurat/T Subset"

project<-"tgfb_lesion_snRNASeq"  

#set up folders
QCfolder = "QC"
Plotfolder = "Plots/T Subset"

#Decide whether to use SCtransform (can take a long time with big datasets; generally recommended)
sct = TRUE

#You can add in housekeeping genes if you want them to be scaled always; otherwise set to "NULL"
add.genes = NULL

#How many genes do you want scaled to use for UMAP/clustering? 
ngenes = 6000

#Which principal components do you want to calculate on? This is a default setting, change if one of the pc's is for something you think is a technical error (i.e. HSP, RP, etc)
pcs = c(1:30)

#Choose what to regress out - could be age, sex, or any metadata column
regress = c("nCount_RNA","percent.mito")

#clustering resolution; the last number will be saved as "seurat_clusters" in metadata
res = c(1.5,1.0,0.5, 0.25, 0.2, 0.15, 0.1, 0.05, 0.02)

#metadata dimensions you want to cluster on
dims = c("seurat_clusters","sample", "subsample", "timepoint","genotype", "microanatomy")

#make a unique name; maybe this is the celltype you've already subset, or the age you're looking at, etc.
iterationname = "t_reclustered"

#Establish cutoffs for heatmaps
pval = 1e-3 #max p-value for significance
lfc = 0.2 #minimum log fold change
minpct = 0.05 #if you want to subset by percent cells in that cluster expressing the gene
maxpct = 1
single = F #should each gene be a marker of a single cluster only
hgenes = 3 #how many genes should be in the heatmap per cluster
ncells = 100 #max # of cells per heatmap column
column = "seurat_clusters" #division you care about
```

Find variable features, normalize, scale, run PCA, clustering, umap
The following is the standard method of normalization and scaling. Interchangeable with the next chunk. Run both, you should have already specified which to use with "sct = T or F". Will take 5-20 minutes to run. 
```{r,results = 'hide',eval = FALSE}
if (!sct){
  sobject.sub.t <- NormalizeData(sobject.sub.t,normalization.method = "LogNormalize", scale.factor = 10000)
  sobject.sub.t<-FindVariableFeatures(sobject.sub.t, selection.method = "vst", nfeatures = ngenes)
  
  all.genes<-rownames(sobject.sub.t)
  var.genes = VariableFeatures(sobject.sub.t)
  add.genes = add.genes[!add.genes %in% var.genes]
  
  any(add.genes %in% var.genes)
  scalegenes = c(var.genes,add.genes)
  VariableFeatures(sobject.sub.t) = scalegenes
  sobject.sub.t<-ScaleData(sobject.sub.t,features = VariableFeatures(sobject.sub.t), vars.to.regress = regress)
}
```

Alternative: SCTransform (great for smaller datasets)
```{r, results='hide',eval = FALSE}
if (sct){
  sobject.sub.t <- SCTransform(sobject.sub.t, vars.to.regress = regress, verbose = FALSE,variable.features.n = ngenes,conserve.memory = T)
}
```

Show most variable genes
```{r}
labels <- c(head(VariableFeatures(sobject.sub.t),10),add.genes)
plot1 = VariableFeaturePlot(sobject.sub.t)
LabelPoints(plot=plot1, points = labels, repel = F, xnudge = 0.1, ynudge = 0.5)
```

Run PCA analysis and show elbow plot
```{r}
sobject.sub.t <- RunPCA(sobject.sub.t,features = VariableFeatures(sobject.sub.t),npcs = 50, verbose = FALSE)
ElbowPlot(sobject.sub.t,ndims = 50, reduction = "pca")
print(sobject.sub.t[["pca"]], dims = 1:20, nfeatures = 5)
```

Once you are satisfied with pc's, run clustering: 
```{r}
sobject.sub.t<-RunUMAP(sobject.sub.t,reduction = "pca",dims = pcs, verbose = F)
sobject.sub.t<-FindNeighbors(sobject.sub.t,dims=pcs,verbose=F)
sobject.sub.t<-FindClusters(sobject.sub.t,verbose=F,resolution = res)
sobject.sub.t$seurat_clusters = sobject.sub.t$SCT_snn_res.0.5
```

Function to print multiple graphs: 
```{r}
PrintSeuratGraph = function(namecard = "a",seurat_object = sobject.sub.t,graphtype = "feature",feature = NULL,group = NULL,split=NULL,cellnames=NULL, label=FALSE){
  if (!is.null(cellnames)){
    #Idents(seurat_object) = cellnames[1]
    #cells = colnames(seurat_object)[Idents(seurat_object) %in% cellnames[1:length(cellnames)]]} 
    cells = cellnames}
  else {cells = cellnames}
  if (graphtype == "feature"){
    graph = FeaturePlot(seurat_object,features = feature,split.by = split, cells = cells,cols = c("lightyellow","darkred"))
  }
  if (graphtype == "violin"){
    graph = VlnPlot(seurat_object,features = feature, pt.size = 0.1, idents = cellnames[2:length(cellnames)],group.by = group, split.by = split)
  }
  if (graphtype == "dim"){
    graph = DimPlot(seurat_object,cells = cells, group.by = group, split.by = split, label = label)
  }
  name = paste0(feature,"_",graphtype,namecard,".eps")
  graph
  setEPS()
  postscript(file.path(dir,Plotfolder,name))
  print(graph)
  dev.off()
}
```
Save the clustered dataset. Overwrite the existing, subset dataset. 
```{r,eval = FALSE}
save(sobject.sub.t,file = file.path(dir,datafolder,paste0(project,"_",iterationname,".RData")))
```

Block to print multiple graphs: 
```{r}
Plotfolder = "Plots/T Subset"
DefaultAssay(sobject.sub.t) = "RNA"

name = paste0(project,"_",iterationname)
genes = c("Cd4", "Cd8a", "Ifng", "Gzmb", "Mki67", "Tcf7", "Il2", "Prf1", "Tbx21", "Gata3", "Rorc", "Itgae")
groups = c("timepoint", "condition", "seurat_clusters", "ttype", "ttypecluster")
genes = genes[genes %in% rownames(GetAssayData(sobject.sub.t,slot = "data"))]

for(feature in genes){
  PrintSeuratGraph(namecard = name,graphtype = "feature",feature = feature)
}

#dim plots for clustering
for(group in groups){
  PrintSeuratGraph(namecard = name,graphtype = "dim",group = group, feature = group)
  PrintSeuratGraph(namecard = paste0(name,"_labeled_"),graphtype = "dim",group = group, feature = group, label = TRUE)
  PrintSeuratGraph(namecard = paste0(name, "_split_"), graphtype = "dim", feature = group, split=group)
}

#violin plots
for(feature in c(genes)){
  PrintSeuratGraph(namecard = name,graphtype = "violin",feature = feature,group = "seurat_clusters")
}

PrintSeuratGraph(namecard = paste0(name, "_timepoint-condition"), graphtype = "dim", group = "timepoint", split = "condition")
PrintSeuratGraph(namecard = paste0(name, "_condition-timepoint"), graphtype = "dim", group = "condition", split = "timepoint")

PrintSeuratGraph(namecard = paste0(name, "_ttype-condition"), graphtype = "dim", group = "ttype", split = "condition")

Plotfolder = "Plots/figure"
Idents(sobject.sub.t) = "condition"
p1 = DimPlot(subset(sobject.sub.t, ident = c("WT", "cKO")), group.by = "condition", split.by = "timepoint", label=F)
p1$data$timepoint = factor(p1$data$timepoint, levels=c("7dpi", "21dpi"))
setEPS()
postscript(file.path(dir, Plotfolder, "condition_timepoint_t.eps"), width = 9, height = 6)
p1
dev.off()
p1 = DimPlot(subset(sobject.sub.t, ident = c("WT", "cKO")), group.by = "condition", label=F)
setEPS()
postscript(file.path(dir, Plotfolder, "condition__t.eps"), width = 6, height = 6)
p1
dev.off()
```

```{r}
#Edit this part carefully. You can add any number of types. Each cluster can only be one type. 

type0 = c(-1)
name0 = "CD8"
type1 = c(0)
name1 = "CD4_Th1"
type2 = c(1)
name2 = "NK"
type3 = c(2)
name3 = "Treg"
type4 = c(3)
name4 = "gdT"
type5 = c(4)
name5 = "T_other"
type6 = c(5)
name6 = "Dead_ribo"
type7 = c(6)
name7 = "Prolif"
type8 = c(7)
name8 = "ILC2"


#Initialize the cluster levels as a vector and replace the cluster levels with the appropriate name. 
clusters = as.factor(sobject.sub.t$seurat_clusters)
type = levels(clusters)
type[type0+2] = name0
type[type1+2] = name1
type[type2+2] = name2
type[type3+2] = name3
type[type4+2] = name4
type[type5+2] = name5
type[type6+2] = name6
type[type7+2] = name7
type[type8+2] = name8

levels(clusters) = type

#Add a metadata column
sobject.sub.t$ttype = clusters

#check the celltype assignment for accuracy
table(sobject.sub.t$ttype,sobject.sub.t$seurat_clusters)

#add a metadata column labelling each cluster
sobject.sub.t$ttypecluster = paste0(sobject.sub.t$ttype,"-",sobject.sub.t$seurat_clusters)
Idents(sobject.sub.t) = "ttype"
```
Save the clustered dataset with fibtype labels. Overwrite the existing, clustered dataset. 
```{r,eval = FALSE}
save(sobject.sub.t,file = file.path(dir,datafolder,paste0(project,"_",iterationname,".RData")))
```


Save the tree
```{r}
# Idents(sobject.sub.fib) = column
Idents(sobject.sub.t) = "seurat_clusters"
sobject.sub.t= BuildClusterTree(sobject.sub.t,dims = pcs)
tree = sobject.sub.t@tools$BuildClusterTree
setEPS()
postscript(file.path(dir,"Trees",paste0(project,"_",iterationname,"_tree.eps")))
plot.phylo(tree, use.edge.length = T, direction = "rightwards")
dev.off()
```


*Note: already scaled/normalized (before subsetting, in 2_Processing.Rmd)*

*Heatmap*

First calculate DE genes for every cluster
```{r,results = "hide",eval = FALSE}
Idents(sobject.sub.t) = column
#markers_all <- FindAllMarkers(
#    object = sobject.sub.immune,
#    assay = "RNA",  ##Edited to do DEG analysis on RNA counts, not SCT transformed data
#    features = rownames(sobject),
#    test.use = "MAST",
#    only.pos = FALSE, 
#    min.pct = 0.05, 
#    logfc.threshold = 0.2
#)

#or

markers_all <- RunPrestoAll(
    object = sobject.sub.t,
    assay = "RNA",
    features = rownames(sobject.sub.t),
    test.use = "wilcox",
    only.pos = FALSE,
    min.pct = 0.05,
    logfc.threshold = 0.2)
write.csv(markers_all,file = file.path(dir,"Spreadsheets",paste0(project, "_", iterationname, "_all_markers.csv")))
```


Make the heatmap
```{r}
#read in a de gene file

markers = read.csv(file.path(dir,"Spreadsheets",paste0(project, "_", iterationname, "_all_markers.csv")),stringsAsFactors = F)

#Select only the genes that pass thresholds
markers = markers[markers$p_val_adj<pval,]

#pick only positives, or restrict by min/max pct expression using pct1/2
markers = markers[markers$avg_log2FC > lfc,] #make log2FC >0.2 for heatmaps, >0 for volcano plots
markers = markers[markers$pct.1 > minpct & markers$pct.2 < maxpct,]

#If you want, select markers that define a single cluster
if (single){markers <- markers[markers$gene %in% names(table(markers$gene))[table(markers$gene) == 1],] }

table(markers$cluster)

topgenes <- markers %>% group_by(cluster) %>% top_n(10, avg_log2FC) #top 10 genes
topgenes = topgenes[order(topgenes$cluster),]

#Subset each cluster to ncells
cellnames = sobject.sub.t@meta.data[,column]
names(cellnames) = colnames(sobject.sub.t)
clusters = levels(as.factor(cellnames))
newcellnames = NULL
for (cluster in clusters){
  n = length(cellnames[cellnames == cluster])
  if (n > ncells){n = ncells}
  newcluster = sample(cellnames[cellnames == cluster],n, replace = F)
  newcellnames = c(newcellnames,newcluster)
}

#check
table(newcellnames)

#Make heatmap
setEPS()
postscript(file.path(dir,"Heatmaps", paste0(project,"_",iterationname, "_", column,"_",pval,"_ncells",ncells,"heatmap.eps")))
DoHeatmap(
    object = sobject.sub.t, 
    features = c(topgenes$gene),
    cells = names(newcellnames),
    group.by = column,
    size = 4,
    label = T,
    draw.lines = T,
    raster = FALSE
) + theme(axis.text.y = element_text(size = 4))

dev.off()

#dotplot for figure
topgenes <- markers %>% group_by(cluster) %>% top_n(3, avg_log2FC) #top 3 genes
topgenes = topgenes[order(topgenes$cluster),]

setEPS()
postscript(file.path(dir,"DotPlots", paste0(project,"_",iterationname,"_","t_dotplot.eps")), width = 8, height = 3.1)
DotPlot(sobject.sub.t, assay = "RNA", features = unique(c(topgenes$gene)), group.by = "ttype", cols = "RdBu") + theme(axis.text.x = element_text(angle = 90, hjust = 0.95))
dev.off()
```

```{r}
#add immunetype to total sobject
#load(file.choose) #load total sobject

celltype = as.character(sobject$celltype)
names(celltype) = as.character(colnames(sobject))

celltypecluster = as.character(sobject$celltypecluster)
names(celltypecluster) = as.character(colnames(sobject))

ttype = as.character(sobject.sub.t$ttype)
names(ttype) = as.character(colnames(sobject.sub.t))

ttypecluster = as.character(sobject.sub.t$ttypecluster)
names(ttypecluster) = as.character(colnames(sobject.sub.t))

for(cell in names(celltype)){
  if(cell %in% names(ttype)){
      index_cell = which(names(celltype) == cell)
      index_immune = which(names(ttype) == cell)
      celltype[index_cell] = paste0("T_", ttype[index_immune])
      celltypecluster[index_cell] = paste0("T_", ttypecluster[index_immune])
  }
}

sobject$ttype = celltype
sobject$ttypecluster = celltypecluster

Idents(sobject) = "ttype"

iterationname = "final_withfibtype_withimmunetype_withttype"

datafolder = "Data/Seurat"
save(sobject,file = file.path(dir,datafolder,paste0(project,"_",iterationname,".RData")))
```
***VOLCANO PLOTS***

***~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~***
TGFb: compare Col1a2 cKO or ADWA11 vs. WT for
-all T clusters together
-each T cluster individually
***~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~***
```{r}
Plotfolder = "Plots/T Subset/Volcano"
groups = c("NULL", "0", "1", "2", "3", "4", "5", "6", "8")
clusters = levels(as.factor(sobject.sub.t$ttypecluster))
#manual assignment
group_names=c("all", clusters[3], clusters[1], clusters[7], clusters[2], clusters[5], clusters[9], clusters[4], clusters[5]) #skipping Prolif-7, fewer than 3 cells
Idents(sobject.sub.t) = "seurat_clusters"

#WT vs. cKO
for(group in groups){
  group_name = group_names[which(groups == group)]
  print(paste0("Finding markers (WT vs. cKO) for: ", group_name))

  markers = FindMarkers(
    sobject.sub.t,
    group.by = "condition",
    ident.1 = "cKO",
    ident.2 = "WT",
    subset.ident = if(group_name == "all"){NULL} else{group}) #either cluster of interest or all clusters together
  
  write.csv(markers,file = file.path(dir,"Spreadsheets/WT_vs_cKO/T",paste0(project, "_", iterationname, "_cluster-", group_name, "_wt-vs-col1a2-cKO_markers.csv")))

  #volcano plot
  
  #Minimum fold change (i.e. 1.15 = 15% increase)
  minfc = 1.15
  #Max adj. p value
  alpha = 0.05
  #Genes to highlight
  ngenes = 100
  
  defile = paste0(project, "_", iterationname, "_cluster-", group_name, "_wt-vs-col1a2-cKO_markers.csv")
  
  #Comparison being made (must match file names)
  comparison = paste0(project, "_", iterationname, "_cluster-", group_name, "_wt-vs-col1a2-cKO_markers")

  newlist=list(de)
  de = read.csv(file.path(dir,"Spreadsheets/WT_vs_cKO/T", defile),stringsAsFactors = F) #any   spreadsheet with gene symbols or other identifiers
  colnames(de)[1] = "Gene"

  #NOTE: generated from "DEG_analysis" script
  fc = de
  fc = fc[!is.na(fc$avg_log2FC),]
  colorkeysdown = fc$Gene[fc$avg_log2FC < -log2(minfc) & fc$p_val_adj < alpha]
  colorkeysup = fc$Gene[fc$avg_log2FC > log2(minfc) & fc$p_val_adj < alpha]
  
  #Either highlight specific genes or pick the top genes in colorkeysup/down
  top = fc[fc$p_val_adj<alpha,]
  top = top[order(top$avg_log2FC),"Gene"]
  highlight_top = c(head(top,ngenes),tail(top,ngenes))
  genelists = list(highlight_top)#, highlight_immune, highlight_fib)
  names_genelists = c("top")#, "immune", "fibrosis")
  size_labels = c(2)#,4, 4)

  allcolors = rep("darkgrey",length(fc$Gene))
  names(allcolors) = fc$Gene
  allcolors[names(allcolors) %in% colorkeysdown] = "blue"
  allcolors[names(allcolors) %in% colorkeysup]= "red"
  # names(allcolors)[allcolors == "yellow"] = "labelled"
  names(allcolors)[allcolors == "red"] = "Col1a2cre Tgfbr2flox"
  names(allcolors)[allcolors == "darkgrey"] = "-"
  names(allcolors)[allcolors == "blue"] = " WT" #space for alphabetizing
    
  #manual label
  clustername = (if(group_name == "all"){} else{paste0("(cluster ", group_name, ") ")})
  name = paste0("T ", clustername, "WT vs. Col1a2 cKO")
  range = max(max(fc$avg_log2FC), min(fc$avg_log2FC))
  if(is.infinite(range)){range = 150}

  for(i in 1:length(names_genelists)){
    setEPS()
    postscript(file.path(dir, Plotfolder, "ctrl_vs_cKO", names_genelists[i], paste0("Volcano_",comparison,"_", names_genelists[i], ".eps")))
    print(EnhancedVolcano(fc,
                lab = fc$Gene,
                x = 'avg_log2FC',
                y = 'p_val_adj',
                xlim = c(-range, range),
                title = name,
                drawConnectors = T,
                arrowheads = F,
                legendPosition = 'bottom',
                pCutoff = alpha,
                FCcutoff = log2(minfc),
                selectLab = genelists[[i]],
                labSize = size_labels[i],
                pointSize = 0.5,
                col=c('black', 'black', 'black', 'red3'),
                colCustom = allcolors,
                gridlines.major = F,
                gridlines.minor = F,
                colAlpha = 1,
                max.overlaps = 500))
     dev.off()
  }
}

#Ctrl vs. ADWA11
for(group in groups){
  group_name = group_names[which(groups == group)]
  print(paste0("Finding markers (ctrl vs. ADWA11) for: ", group_name))

  markers = FindMarkers(
    sobject.sub.t,
    group.by = "condition",
    ident.1 = "ADWA11",
    ident.2 = "WT",
    subset.ident = if(group_name == "all"){NULL} else{group}) #either cluster of interest or all clusters together
  
  write.csv(markers,file = file.path(dir,"Spreadsheets/WT_vs_ADWA11/T",paste0(project, "_", iterationname, "_cluster-", group_name, "_ctrl-vs-adwa11_markers.csv")))

  #volcano plot
  
  #Minimum fold change (i.e. 1.15 = 15% increase)
  minfc = 1.15
  #Max adj. p value
  alpha = 0.05
  #Genes to highlight
  ngenes = 100
  
  defile = paste0(project, "_", iterationname, "_cluster-", group_name, "_ctrl-vs-adwa11_markers.csv")
  
  #Comparison being made (must match file names)
  comparison = paste0(project, "_", iterationname, "_cluster-", group_name, "_ctrl-vs-adwa11_markers")

  newlist=list(de)
  de = read.csv(file.path(dir,"Spreadsheets/WT_vs_ADWA11/T", defile),stringsAsFactors = F) #any   spreadsheet with gene symbols or other identifiers
  colnames(de)[1] = "Gene"

  #NOTE: generated from "DEG_analysis" script
  fc = de
  fc = fc[!is.na(fc$avg_log2FC),]
  colorkeysdown = fc$Gene[fc$avg_log2FC < -log2(minfc) & fc$p_val_adj < alpha]
  colorkeysup = fc$Gene[fc$avg_log2FC > log2(minfc) & fc$p_val_adj < alpha]
  
  #Either highlight specific genes or pick the top genes in colorkeysup/down
  top = fc[fc$p_val_adj<alpha,]
  top = top[order(top$avg_log2FC),"Gene"]
  highlight_top = c(head(top,ngenes),tail(top,ngenes))
  genelists = list(highlight_top)#, highlight_immune, highlight_fib)
  names_genelists = c("top")#, "immune", "fibrosis")
  size_labels = c(2)#,4, 4)

  allcolors = rep("darkgrey",length(fc$Gene))
  names(allcolors) = fc$Gene
  allcolors[names(allcolors) %in% colorkeysdown] = "blue"
  allcolors[names(allcolors) %in% colorkeysup]= "red"
  # names(allcolors)[allcolors == "yellow"] = "labelled"
  names(allcolors)[allcolors == "red"] = "ADWA11"
  names(allcolors)[allcolors == "darkgrey"] = "-"
  names(allcolors)[allcolors == "blue"] = " WT" #space for alphabetizing
    
  #manual label
  clustername = (if(group_name == "all"){} else{paste0("(cluster ", group_name, ") ")})
  name = paste0("T ", clustername, "Ctrl vs. ADWA11")
  range = max(max(fc$avg_log2FC), min(fc$avg_log2FC))
  if(is.infinite(range)){range = 150}

  for(i in 1:length(names_genelists)){
    setEPS()
    postscript(file.path(dir, Plotfolder, "ctrl_vs_adwa11", names_genelists[i], paste0("Volcano_",comparison,"_", names_genelists[i], ".eps")))
    print(EnhancedVolcano(fc,
                lab = fc$Gene,
                x = 'avg_log2FC',
                y = 'p_val_adj',
                xlim = c(-range, range),
                title = name,
                drawConnectors = T,
                arrowheads = F,
                legendPosition = 'bottom',
                pCutoff = alpha,
                FCcutoff = log2(minfc),
                selectLab = genelists[[i]],
                labSize = size_labels[i],
                pointSize = 0.5,
                col=c('black', 'black', 'black', 'red3'),
                colCustom = allcolors,
                gridlines.major = F,
                gridlines.minor = F,
                colAlpha = 1,
                max.overlaps = 500))
     dev.off()
  }
}

```

```{r}
Plotfolder = "Plots/T Subset/comparison"

Idents(sobject.sub.t) = "condition"
compare = subset(sobject.sub.t, ident = c("WT", "cKO"))

setEPS()
postscript(file.path(dir,Plotfolder, "itgae.eps"))
VlnPlot(compare, "Itgae", split.by = "condition")
dev.off()

markers = c("Itgae", "Runx2", "Runx3")
compare <- AddModuleScore(
    compare,
    markers,
    pool = NULL,
    nbin = 24,
    ctrl = 100,
    name = paste0("trm_score"),
    seed = 1,
    search = TRUE)


my.pal = palette(brewer.pal(n = 8, name = "YlOrRd"))
compare$condition = factor(compare$condition, levels = c("WT", "cKO"))

p1 = Plot_Density_Custom(compare, c("Itgae", "Runx2", "Runx3"), joint=T, custom_palette = my.pal) + facet_grid(.~compare$condition)
setEPS()
postscript(file.path(dir,Plotfolder, "trm.eps"), width = 24, height = 15)
p1
dev.off()

postscript(file.path(dir,Plotfolder, "trm_vln.eps"))
VlnPlot(compare, "trm_score1")
dev.off()
```

```{r}
compare_avg <- AverageExpression(compare, return.seurat = TRUE, verbose = TRUE)
markers = c("Itgae", "Cd69",  "Ifng", "Zfp683", "Prdm1", "Runx3", "Tcf7", "Ccr2")

DoHeatmap(
    object = compare_avg, 
    features = markers,
    size = 5,
    label = T,
    raster = F,
    draw.lines = F) + scale_fill_gradientn(colors = rev(RColorBrewer::brewer.pal(n =4, name = "RdBu"))) + theme(axis.text.y = element_text(size = 6))

VlnPlot(subset(compare, ident = "CD8"), group.by = "condition", markers, stack=T)
```

