---
title: "Bourne Resting"
author: "Nathan Ewing-Crystal"
date: "7/28/2022"
output: html_document
---

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r}
#install.packages('Seurat')
library(Seurat)
#install.packages('ggplot2')
library(ggplot2)
#install.packages('dplyr')
library(dplyr)
#install.packages('ape')
library(ape)
#install.packages('cowplot')
library(cowplot)
#install.packages('Matrix')
library(Matrix)
#install.packages('EnhancedVolcano')
library(EnhancedVolcano)
#install.packages('knitr')
library(knitr)
#install.packages('readr')
library(readr)
#if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
#BiocManager::install("MAST")
library(MAST)
#install.packages('gplots')
library(gplots)
```

*Edit the following code every time* 
and make sure the folders "QC" and "Plots" and "Heatmaps" and "Trees" and "Spreadsheets" and "Data/Seurat" are present in the "dir" folder
```{r}
#Specify your preferred directory for all input + output
dir= "/Users/nathanewing-crystal/Ari Molofsky Lab Dropbox/Ari Molofsky Lab Team Folder/Nathan Ewing-Crystal/RNASeq/Bourne Astrocytes/Resting (Bourne Lab)"
# dir = "/mnt/DATA/molofsky_lab/newing-crystal/Bourne Astrocytes/Resting (Bourne Lab)"
setwd(dir)

#Specify exactly where your seurat files live
datafolder = "Data"

#This name needs to match your project name within the seurat object
project<-"resting-marmoset" 

#set up folders
#QCfolder = "QC"
Plotfolder = "Plots"

#Important genes to determine your cells of interest
igenes = c("COL1A1", "PDGFRA", "PDGFRB", "DCN", "TGFBR2", "ITGB8", "ITGB1", "ITGAV", "OLIG2", "GFAP", "PTPRC", "ITGAM", "DES")

#make a unique name; maybe this is the celltype you've already subset, or the age you're looking at, etc.
iterationname = "v1"

#Which principal components do you want to calculate on? This is a default setting, change if one of the pc's is for something you think is a technical error (i.e. HSP, RP, etc)
pcs = c(1:30)

#clustering resolution; the last number will be saved as "seurat_clusters" in metadata
res = c(1.5,1.0,0.5)

#Establish cutoffs for heatmaps
pval = 1e-3 #max p-value for significance
lfc = 0.2 #minimum log fold change
minpct = 0.05 #if you want to subset by percent cells in that cluster expressing the gene
maxpct = 1
single = F #should each gene be a marker of a single cluster only
hgenes = 3 #how many genes should be in the heatmap per cluster
ncells = 100 #max # of cells per heatmap column
#column = "sub.cluster" #division you care about

```


********BELOW CODE PERFORMED ON PRESIDIO SERVER********

  
```{r}
filename <- file.path(dir, "YA all CTs.rds")
sobj <- readRDS(filename)
```


```{r}
DefaultAssay(sobj) <- "RNA"
FeaturePlot(sobj, "COL1A1")
FeaturePlot(sobj, "CD3E")
DimPlot(sobj)
```

Function to print multiple graphs: 
```{r}
PrintSeuratGraph = function(namecard = "a",seurat_object = sobj,graphtype = "feature",feature = NULL,group = NULL,split=NULL,cellnames=NULL, label=FALSE){
  if (!is.null(cellnames)){
    Idents(seurat_object) = cellnames[1]
    cells = colnames(seurat_object)[Idents(seurat_object) %in% cellnames[2:length(cellnames)]]} 
  else {cells = cellnames}
  if (graphtype == "feature"){
    graph = FeaturePlot(seurat_object,features = feature,split.by = split, cells = cells,cols = c("lightyellow","darkred"))
  }
  if (graphtype == "violin"){
    graph = VlnPlot(seurat_object,features = feature, pt.size = 0.1, idents = cellnames[2:length(cellnames)],group.by = group, split.by = split)
  }
  if (graphtype == "dim"){
    graph = DimPlot(seurat_object,cells = cells, group.by = group, split.by = split, label = label)
    
  }
  name = paste0(feature,"_",graphtype,namecard,".eps")
  graph
  setEPS()
  postscript(file.path(dir,Plotfolder,name))
  print(graph)
  dev.off()
}
```

Block to print multiple graphs: 
```{r}
name = paste0(project,iterationname)
genes = igenes
genes = genes[genes %in% rownames(GetAssayData(sobj,slot = "data"))]

for(feature in genes){
  PrintSeuratGraph(namecard = name,graphtype = "feature",feature = feature)
}


#dim plots for clustering
PrintSeuratGraph(namecard = name,graphtype = "dim")

#violin plots
for(feature in genes){
  PrintSeuratGraph(namecard = name,graphtype = "violin",feature = feature,group = "seurat_clusters")
}
```


```{r}
# markers_all = FindAllMarkers(
#   object = sobj,
#   features = rownames(sobj),
#   test.use = "MAST", 
#   only.pos = FALSE, 
#   min.pct = 0.05, 
#   logfc.threshold = 0.2)

markers_resting_fib = FindMarkers(
  object = sobj,
  ident.1 = "Fibroblast",
  features = rownames(sobj),
  test.use = "MAST",
  only.pos = FALSE,
  min.pct = 0.05,
  logfc.threshold = 0.2)

write.csv(markers_resting_fib, file.path(dir, "/Tables/markers_resting_fib.csv"))
```

***INTEGRATE RESTING FIBS WITH STROKE FIBS***
(https://satijalab.org/seurat/archive/v3.0/integration.html)

```{r}
# Resting fibs: subset on "fibroblast" identity (pre-annotated by Leon Teo)
sobj_rest = subset(sobj, idents = "Fibroblast")

# Stroke fibs: "astrocyte-stroke_subclustered-fib-only_v1.RData" (see processing script for stroke data)
# note - currently named "sub18"
load("/mnt/DATA/molofsky_lab/newing-crystal/Bourne Astrocytes/Data/astrocyte-stroke_subclustered-fib-only_v1.RData")
#rename stroke fibs object
sobj_stroke <- sub18
rm(sub18)

#store cortex vs. meninges in metadata
sobj_stroke$condition <- "7dpi"
sobj_rest$condition <- "Rest"

datasets <- list(sobj_stroke, sobj_rest)

fib.anchors <- FindIntegrationAnchors(object.list = datasets, dims = 1:30, verbose = TRUE)
sobj_integrated <- IntegrateData(anchorset = fib.anchors, dims = 1:30)
```

```{r}
save(sobj_integrated,file = file.path(dir, datafolder,paste0("integrated_fibroblasts_rest_7dpi.RData")))
```

***SWITCH TO LOCAL***

```{r}
load("/Users/nathanewing-crystal/Dropbox (Ari Molofsky Lab)/Ari Molofsky Lab Team Folder/Nathan Ewing-Crystal/RNASeq/Bourne Astrocytes/Resting (Bourne Lab)/Data/integrated_fibroblasts_rest_7dpi.RData")
sobject <- sobj_integrated
```

Run PCA analysis and show elbow plot
```{r}
sobject <- ScaleData(sobject)
sobject <- RunPCA(sobject,features = VariableFeatures(sobject),npcs = 50, verbose = FALSE)
ElbowPlot(sobject,ndims = 50, reduction = "pca")
print(sobject[["pca"]], dims = 1:20, nfeatures = 5)
```

Once you are satisfied with pc's, run clustering: 
```{r}
sobject<-RunUMAP(sobject,reduction = "pca",dims = pcs, verbose = F)
sobject<-FindNeighbors(sobject,dims=pcs,verbose=F)
sobject<-FindClusters(sobject,verbose=F,resolution = 0.2)
```



```{r}
Plotfolder = "Plots/Integrated rest 7dpi"

#dim plots for clustering
PrintSeuratGraph(sobject, namecard = "integrated", graphtype = "dim", label = TRUE)
PrintSeuratGraph(sobject, namecard = "integrated_group", graphtype = "dim", group = "condition")
PrintSeuratGraph(sobject, namecard = "integrated_split", graphtype = "dim", split = "condition")

name = paste0(project, iterationname, "umap_large_labels.eps")
setEPS()
postscript(file.path(dir, Plotfolder, name))
DimPlot(sobject, label = T, label.size = 7, repel=T)
dev.off()

DefaultAssay(sobject) = "RNA"

markers_fibs_integrated = FindAllMarkers(
  object = sobject,
  test.use = "MAST",
  only.pos = FALSE,
  min.pct = 0.05,
  logfc.threshold = 0.2)

write.csv(markers_fibs_integrated, file.path(dir, "/Tables/markers_integrated_resting_7dpi_fib.csv"))
```

```{r}
sobject <- SetIdent(sobject, value = "condition")

markers_rest_vs_7dpi_fib = FindMarkers(
  object = sobject,
  ident.1 = "7dpi",
  ident.2 = "Rest",
  features = rownames(sobject),
  test.use = "MAST",
  only.pos = FALSE,
  min.pct = 0.05,
  logfc.threshold = 0.2)

write.csv(markers_rest_vs_7dpi_fib, file.path(dir, "/Tables/markers_7dpi_vs_resting_fib.csv"))
```

#Heatmap with fibrosis-related genes
```{r}
Plotfolder = "Heatmaps"
name = paste0(project, "_", iterationname)

avgexp <- AverageExpression(sobject, return.seurat = TRUE, verbose = TRUE)
fibtypecluster_new = c(1, 4, 0, 2, 3, 5, 6) #arrange with myofibroblast clusters first
levels(avgexp) = fibtypecluster_new
markers = c("ACTA2", "MKI67", "FBN2", "LUM", "CTHRC1", "POSTN", "FN1", "FAP", "COL1A1", "COL1A2", "DCN")

#sort so that genes are arranged by cluster with maximum expression, in order of desired clusters
df = data.frame(avgexp@assays$RNA@scale.data[markers,])
df_col_arrange = df[,fibtypecluster_new]
markers_max = apply(df_col_arrange, 1, which.max)
df = data.frame(markers, markers_max)
df.sort = arrange(df, markers_max)
markers_sorted = df.sort$markers

setEPS()
postscript(file.path(dir,Plotfolder, paste0(name, "fibro_markers_avg.eps")), height = 6)
DoHeatmap(
    object = avgexp, 
    features = markers_sorted,
    size = 10,
    label = T,
    raster = F,
    draw.lines = F) + scale_fill_gradientn(colors = rev(RColorBrewer::brewer.pal(n =4, name = "RdBu"))) + theme(axis.text.y = element_text(size = 20))
dev.off()
```
Barplot for any two (or more) categories

```{r}
Plotfolder = "Plots/Integrated rest 7dpi"
  
#Pick metadata columns
clustercolumn = "seurat_clusters"
samplecolumn = "condition"
#pick a reasonable number of cells per sample to normalize by
ncells = 2000 
cols = c("red","blue")
name = "_cluster-frequencies"

#Make a table and normalize
r = table(sobject[[clustercolumn]][,1],sobject[[samplecolumn]][,1])

  t = r  
  #remove any clusters that don't have cells
  t = t[rowSums(t)>0,]
  
  #normalize by sample
  t = apply(t,MARGIN = 2,function(x)x/sum(x))
  t = round(t*ncells,0)
  
  #convert to percents for each cluster
  t = apply(t,MARGIN = 1,function(x)x/sum(x))
  t = round(t*100,2)
  
  setEPS()
  postscript(file.path(dir,Plotfolder,paste0(name,"barplot.eps")))
  barplot(t, main="Cluster composition",
       # xlab="Cluster", 
       ylab = "% of cluster", ylim = c(0,100), col=cols,axisnames = T,
        width = .39,xlim = c(0,5), space = 0.6,cex.names = 0.7,axis.lty = 1, las=2, mgp=c())
  title(xlab = "Cluster", line = 4)
  legend("topright", legend=rownames(t), fill=cols, bg="white")
  dev.off()

#Instead of % cluster, % of total cells in each condition
  
  t = r  
  #remove any clusters that don't have cells
  t = t[rowSums(t)>0,]
  
  #normalize by sample
  t = apply(t,MARGIN = 2,function(x)x/sum(x))
  t = round(t*ncells,0)
  
  #convert to percents for each *condition*
  t = apply(t,MARGIN = 2,function(x)x/sum(x))
  t = round(t*100,2)
  
  name = "_condition-frequencies"
  
  setEPS()
  postscript(file.path(dir,Plotfolder,paste0(name,"barplot.eps")), height = 8, width = 5)
  barplot(t(t), main="Condition composition", beside = T,
       ylab = "% of condition", col=cols,axisnames = T, ylim=c(0,70),
        width = .2, cex.names = 0.7,axis.lty = 1, las=2, mgp=c())
  title(xlab = "Cluster", line = 4)
  legend("topright", legend=rownames(t(t)), fill=cols, bg="white", inset=c(0.08, 0.01))
  dev.off()
```

```{r}
save(sobject,file = file.path(dir, datafolder, "integrated_fibroblasts_rest_7dpi.RData"))
```

```{r}
#Make myofibroblast score

require(biomaRt)

features_tgfb_filter = read.csv("/Users/nathanewing-crystal/Ari Molofsky Lab Dropbox/Ari Molofsky Lab Team Folder/Nathan Ewing-Crystal/RNASeq/Visium/Seurat/Data/Other/tgfb-vs-pbs.csv")

features_tgfb_filter = features_tgfb_filter$Cs

### If you are using a mouse data, then its needed to convert the gene names to human orthologs

marmoset = useMart("ensembl", dataset = "cjacchus_gene_ensembl", host = "dec2021.archive.ensembl.org")
mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl", host = "dec2021.archive.ensembl.org")

#change to rownames(matrix1) from rownames(alldata@assays$RNA@data)
genesV2 = getLDS(attributes = c("mgi_symbol"), filters = "mgi_symbol", values = features_tgfb_filter , mart = mouse, attributesL = c("hgnc_symbol","hgnc_id","ensembl_gene_id", "external_gene_name"), martL = marmoset, uniqueRows=T)
print(head(genesV2))
features_tgfb_filter_marmoset = genesV2$Gene.name
features_tgfb_filter_marmoset_v2 = features_tgfb_filter_marmoset[!(features_tgfb_filter_marmoset == "")]
genes = list(features_tgfb_filter_marmoset_v2)

sobject <- AddModuleScore(
  sobject,
  genes,
  pool = NULL,
  nbin = 24,
  ctrl = 100,
  name = "tgfb_score",
  seed = 1,
  search = TRUE,
)

Idents(sobject) = "timepoint"
Idents(sobject) = factor(sobject$condition, levels = c("Rest", "7dpi"))

Plotfolder = "/Users/nathanewing-crystal/Ari Molofsky Lab Dropbox/Ari Molofsky Lab Team Folder/Nathan Ewing-Crystal/RNASeq/Bourne Astrocytes/Resting (Bourne Lab)/Plots/Integrated rest 7dpi/Myofib_score"
name = "tgfb_score.eps"

  setEPS()
  postscript(file.path(Plotfolder,name))
VlnPlot(sobject, "tgfb_score1")
  dev.off()
  
    #sig
sobject_metadata = data.frame(sobject@meta.data) #creates dataframe of metadata in sobject
wilcox.test(sobject_metadata$tgfb_score1 ~ sobject_metadata$condition)
```
